From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Fri, 25 Jul 2025 16:28:47 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index bf2f7c4959f9f2cd43bbd4d4310e7eedf10cf892..6eadc3ca6b7b4d5ae61dd9bc7b564db63e495dfc 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4660,6 +4660,34 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public Collection<GeneratedStructure> getStructures(int x, int z, @NotNull Structure structure);
 
+    /**
+     * Submit a task to be executed on the world thread.
+     * Usually after all other thread tasks have been executed.
+     * @param task the task to execute
+     */
+    void submitCallbackTask(@NotNull Runnable task);
+
+    /**
+     * Submit a task to be executed on the world thread.
+     * Usually after all this world thread tasks have been executed.
+     * @param task the task to execute
+     */
+    void submitScopedTask(@NotNull Runnable task);
+
+    /**
+     * Submit a task to be executed on the world thread every tick.
+     * @param id
+     * @param task
+     */
+    void submitCyclicalTask(@NotNull String id, @NotNull Runnable task);
+
+    /**
+     * Removes a cyclical task
+     * @param id
+     * @return
+     */
+    @Nullable Runnable removeCyclicalTask(@NotNull String id);
+
     /**
      * Represents various map environment types that a world may be
      */
