From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Fri, 25 Jul 2025 16:28:47 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 0d39006698a60f7dbf47892e80c2d3b0c9f4f027..2e5296f2648153ccbf0f0147a5d42d61ded2dfd3 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4686,6 +4686,34 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public Collection<GeneratedStructure> getStructures(int x, int z, @NotNull Structure structure);
 
+    /**
+     * Submit a task to be executed on the world thread.
+     * Usually after all other thread tasks have been executed.
+     * @param task the task to execute
+     */
+    void submitCallbackTask(@NotNull Runnable task);
+
+    /**
+     * Submit a task to be executed on the world thread.
+     * Usually after all this world thread tasks have been executed.
+     * @param task the task to execute
+     */
+    void submitScopedTask(@NotNull Runnable task);
+
+    /**
+     * Submit a task to be executed on the world thread every tick.
+     * @param id
+     * @param task
+     */
+    void submitCyclicalTask(@NotNull String id, @NotNull Runnable task);
+
+    /**
+     * Removes a cyclical task
+     * @param id
+     * @return
+     */
+    @Nullable Runnable removeCyclicalTask(@NotNull String id);
+
     /**
      * Represents various map environment types that a world may be
      */
