From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Thu, 24 Jul 2025 11:18:56 +0800
Subject: [PATCH] Rebrand to Octopus


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 4e099b584c3435ba1ebce492011d35ae1ac0cdf8..6c428744c2045248725ee35f8bc2b5e04716c329 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -16,6 +16,7 @@ public interface ServerBuildInfo {
      * The brand id for Paper.
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
+    Key BRAND_OCTOPUS_ID = Key.key("happyland", "octopus");
 
     /**
      * Gets the {@code ServerBuildInfo}.
