From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 23:11:43 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 48ba9df053e6266cb651e3c7c93cc5ba771d48d1..d85ccad29a378c771163d4c1cbd0118959419967 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -104,6 +104,16 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     void setVoidDamageMinBuildHeightOffset(double minBuildHeightOffset);
     // Paper end - void damage configuration
 
+    // Octopus start - Get the MSPT for a single world
+    /**
+     * Get the mspt values from the last 5 seconds。
+     *
+     * @return a collection of mspt values.
+     */
+    @NotNull
+    public long[] getTickTimes();
+    // Octopus end - Get the MSPT for a single world
+
     // Paper start
     /**
      * @return The amount of entities in this world
