From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 23:11:54 +0800
Subject: [PATCH] Add clientConnected API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 190313bc060d5f8440cec7c249ad1d2f1074f5bc..e260e71f04ab1458feb4e0d3aa4d1e13c10f345a 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3614,6 +3614,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         builder.hat(this.getClientOption(ClientOption.SKIN_PARTS).hasHatsEnabled());
     }
 
+    // Octopus start
+    /**
+     * Gets whether the player is connected to the server.
+     * @return
+     */
+    boolean clientConnected();
+    // Octopus end
+
     // Paper start - Player Profile API
     /**
      * Gets a copy of this players profile
