From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 23:11:54 +0800
Subject: [PATCH] Add clientConnected API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fe3f4b7552e1230d518a8554a3f0fd4009523a73..61c1f6dc2613817f5510bbfd78d48a00b4679e62 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3568,6 +3568,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         builder.hat(this.getClientOption(ClientOption.SKIN_PARTS).hasHatsEnabled());
     }
 
+    // Octopus start
+    /**
+     * Gets whether the player is connected to the server.
+     * @return
+     */
+    boolean clientConnected();
+    // Octopus end
+
     // Paper start - Player Profile API
     /**
      * Gets a copy of this players profile
