From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sun, 23 Feb 2025 20:51:33 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index dfe3b20f16fbd45f71206fa95dd03c374b44aaf6..c54068a961fb83925bbfd063d6097489007d349f 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -101,6 +101,16 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     void setVoidDamageMinBuildHeightOffset(double minBuildHeightOffset);
     // Paper end - void damage configuration
 
+    // Octopus start - Get the MSPT for a single world
+    /**
+     * Get the mspt values from the last 5 seconds。
+     *
+     * @return a collection of mspt values.
+     */
+    @NotNull
+    public long[] getTickTimes();
+    // Octopus end - Get the MSPT for a single world
+
     // Paper start
     /**
      * @return The amount of Entities in this world
