From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Thu, 1 May 2025 12:18:11 +0800
Subject: [PATCH] Add LoadOrder ROOT


diff --git a/src/main/java/org/bukkit/plugin/PluginLoadOrder.java b/src/main/java/org/bukkit/plugin/PluginLoadOrder.java
index b77436fdda114cc28cd9cc7d49c537afeb8f26ea..5a4cb045c6ea3d3b2c7590337d405620fcdc3d79 100644
--- a/src/main/java/org/bukkit/plugin/PluginLoadOrder.java
+++ b/src/main/java/org/bukkit/plugin/PluginLoadOrder.java
@@ -13,5 +13,9 @@ public enum PluginLoadOrder {
      * Indicates that the plugin will be loaded after the first/default world
      * was created
      */
-    POSTWORLD
+    POSTWORLD,
+    /**
+     * Indicates that the plugin will be loaded before all other plugins
+     */
+    ROOT
 }
