--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.entity.item;
 
 import com.mojang.logging.LogUtils;
+
+import java.util.function.Consumer;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.CrashReportCategory;
@@ -396,12 +_,15 @@
 
     @Nullable
     @Override
-    public Entity teleport(TeleportTransition teleportTransition) {
+    public Entity teleport(TeleportTransition teleportTransition, Consumer<Entity> callback) { // Octopus
         ResourceKey<Level> resourceKey = teleportTransition.newLevel().dimension();
         ResourceKey<Level> resourceKey1 = this.level().dimension();
         boolean flag = (resourceKey1 == Level.END || resourceKey == Level.END) && resourceKey1 != resourceKey;
-        Entity entity = super.teleport(teleportTransition);
-        this.forceTickAfterTeleportToDuplicate = entity != null && flag && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowUnsafeEndPortalTeleportation; // Paper
-        return entity;
+        // Paper
+        return super.teleport(teleportTransition, teleported -> { // Octopus
+            this.forceTickAfterTeleportToDuplicate = teleported != null && flag && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowUnsafeEndPortalTeleportation; // Paper
+
+            if (callback != null) callback.accept(teleported);
+        }); // Octopus
     }
 }
