--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.item;
 
 import java.util.Optional;
+import java.util.function.Consumer;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
@@ -231,13 +_,18 @@
 
     @Nullable
     @Override
-    public Entity teleport(TeleportTransition teleportTransition) {
-        Entity entity = super.teleport(teleportTransition);
-        if (entity instanceof PrimedTnt primedTnt) {
-            primedTnt.setUsedPortal(true);
-        }
+    public Entity teleport(TeleportTransition teleportTransition, Consumer<Entity> callback) { // Octopus
+        // Octopus start
+        return super.teleport(teleportTransition, teleported -> {
+            if (teleported != null) {
+                if (teleported instanceof PrimedTnt primedTnt) {
+                    primedTnt.setUsedPortal(true);
+                }
+            }
 
-        return entity;
+            if (callback != null) callback.accept(teleported);
+        });
+        // Octopus end
     }
 
     @Override
