From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Fri, 25 Jul 2025 00:02:52 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 7dda83ba3b28c169ff7c2dfc37210b7d4474cdd3..0926a883a5a9ac604fbe94dd9db21005cc42691c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -211,6 +211,13 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     }
     // Paper end - void damage configuration
 
+    // Octopus start - Get the MSPT for a single world
+    @Override
+    public @NotNull long[] getTickTimes() {
+        return world.getTickTimes();
+    }
+    // Octopus end - Get the MSPT for a single world
+
     // Paper start - Provide fast information methods
     @Override
     public int getEntityCount() {
