From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Wed, 31 Dec 2025 18:53:46 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
index 3bb9b58ee97464687e348e23b99159226415c267..dd5a6829bc1466c6741f7b66db2fef9824306eb5 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
@@ -181,6 +181,12 @@ public final class MoonriseRegionFileIO {
         long failures = 1L; // start at 0.13ms
 
         while (taskController.hasTasks()) {
+            // Octopus start - Parallel world ticking
+            if (pollWorldTasks(world)) {
+                failures = 1L;
+                continue;
+            }
+            // Octopus end - Parallel world ticking
             Thread.yield();
             failures = ConcurrentUtil.linearLongBackoff(failures, 125_000L, 5_000_000L); // 125us, 5ms
         }
@@ -194,6 +200,12 @@ public final class MoonriseRegionFileIO {
             }
 
             if (totalTasks > (long)tasksRemaining) {
+                // Octopus start - Parallel world ticking
+                if (pollWorldTasks(world)) {
+                    failures = 1L;
+                    continue;
+                }
+                // Octopus end - Parallel world ticking
                 Thread.yield();
                 failures = ConcurrentUtil.linearLongBackoff(failures, 125_000L, 5_000_000L); // 125us, 5ms
             } else {
@@ -202,6 +214,12 @@ public final class MoonriseRegionFileIO {
         }
     }
 
+    // Octopus start - Parallel world ticking
+    private static boolean pollWorldTasks(final ServerLevel world) {
+        return TickThread.isTickThread() && world.worldTaskQueueInstance.pollTasks();
+    }
+    // Octopus end - Parallel world ticking
+
     /**
      * Returns the priority associated with blocking I/O based on the current thread. The goal is to avoid
      * dumb plugins from taking away priority from threads we consider crucial.
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index 640548d78b87653871dd1f7b0f84574c5e205fe8..135d593c89312abaa391e60971fa9fc168fd031d 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -553,7 +553,7 @@ public final class ChunkTaskScheduler {
 
         // we could do a simple spinwait here, since we do not need to process tasks while performing this load
         // but we process tasks only because it's a better use of the time spent
-        this.world.getChunkSource().mainThreadProcessor.managedBlock(() -> {
+        this.world.worldTaskQueueInstance.spinWait(() -> { // Octopus
             return ((ChunkSystemServerLevel)this.world).moonrise$getSpecificChunkIfLoaded(chunkX, chunkZ, status) != null;
         });
 
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 162d7890f204315b0476467e757de5c9dd84aeca..161ad24648c7e0ea7f9e608bf498361cbdf0973d 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -307,8 +307,8 @@ public interface DispenseItemBehavior {
                     capturedBlockStates = new java.util.ArrayList<>(level.capturedBlockStates.values());
                     level.capturedBlockStates.clear();
 
-                    treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
-                    net.minecraft.world.level.block.SaplingBlock.treeType = null;
+                    treeType = net.minecraft.world.level.block.SaplingBlock.treeTypeRT.get(); // Octopus
+                    net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(null); // Octopus
                 }
                 // Paper end
                 // CraftBukkit start
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 1049f78b1300a584f87579ee5af83145f46dac18..d812240f7b93c0f591596bcd0067a8c43ee5222c 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1,5 +1,9 @@
 package net.minecraft.server;
 
+// Octopus start
+import ca.spottedleaf.moonrise.common.util.TickThread;
+import org.apache.commons.lang3.tuple.Pair;
+// Octopus end
 import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
@@ -448,10 +452,14 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             if (currTime - ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getLastMidTickFailure() <= TASK_EXECUTION_FAILURE_BACKOFF) {
                 continue;
             }
-            if (!world.getChunkSource().pollTask()) {
+            if (!world.worldTaskQueueInstance.isOwnedByCurrentThread()) { // Octopus
                 // we need to back off if this fails
                 ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$setLastMidTickFailure(currTime);
             } else {
+                if (!world.worldTaskQueueInstance.pollTasks()) { // Octopus
+                    ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$setLastMidTickFailure(currTime); // Octopus
+                } // Octopus
+
                 executed = true;
             }
         }
@@ -1010,6 +1018,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             if (this.hasStopped) return;
             this.hasStopped = true;
         }
+        this.worldTickExecutor.shutdownNow(); // Octopus
         if (!this.hasLoggedStop && this.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
         this.shutdownThread = Thread.currentThread(); // Paper - Improved watchdog support
         org.spigotmc.WatchdogThread.doStop(); // Paper - Improved watchdog support
@@ -1518,7 +1527,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         boolean ret = false; // Paper - force execution of all worlds, do not just bias the first
         if (this.tickRateManager.isSprinting() || this.shouldRunAllTasks() || this.haveTime()) {
             for (ServerLevel level : this.getAllLevels()) {
-                if (level.getChunkSource().pollTask()) {
+                if (level.getChunkSource().pollTask() || level.worldTaskQueueInstance.pollTasks()) { // Octopus
                     ret = true; // Paper - force execution of all worlds, do not just bias the first
                 }
             }
@@ -1760,6 +1769,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     public final io.papermc.paper.threadedregions.EntityScheduler.EntitySchedulerTickList entitySchedulerTickList = new io.papermc.paper.threadedregions.EntityScheduler.EntitySchedulerTickList(); // Paper - optimise Folia entity scheduler
 
+    // Octopus start
+    private final java.util.concurrent.atomic.AtomicInteger threadIdGenerator = new java.util.concurrent.atomic.AtomicInteger(0);
+    private final java.util.concurrent.ExecutorService worldTickExecutor = java.util.concurrent.Executors.newCachedThreadPool(task -> {
+        final TickThread wrapped = new TickThread(task, "Octopus World Tick Worker - " + threadIdGenerator.getAndIncrement());
+        wrapped.setPriority(6);
+        wrapped.setContextClassLoader(MinecraftServer.class.getClassLoader());
+        return wrapped;
+    });
+    // Octopus end
+
     protected void tickChildren(final BooleanSupplier haveTime) {
         ProfilerFiller profiler = Profiler.get();
         this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.suspendFlushing());
@@ -1810,26 +1829,62 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.updateEffectiveRespawnData();
 
         this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked
+        final java.util.Deque<me.xiaozhangup.octopus.WorldTickFutureTask> tasks = new java.util.ArrayDeque<>(); // Octopus
         for (ServerLevel level : this.getAllLevels()) {
             level.hasPhysicsEvent = org.bukkit.event.block.BlockPhysicsEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - BlockPhysicsEvent
             level.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
             level.updateLagCompensationTick(); // Paper - lag compensation
             net.minecraft.world.level.block.entity.HopperBlockEntity.skipHopperEvents = level.paperConfig().hopper.disableMoveEvent || org.bukkit.event.inventory.InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length == 0; // Paper - Perf: Optimize Hoppers
             profiler.push(() -> level + " " + level.dimension().identifier());
-            profiler.push("tick");
 
-            try {
-                level.tick(haveTime);
-            } catch (Throwable t) {
-                CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
-                level.fillReportDetails(report);
-                throw new ReportedException(report);
-            }
+            // Octopus start
+            java.util.concurrent.Callable<Pair<ServerLevel, Throwable>> tickingFunction = () -> {
+                level.worldTaskQueueInstance.acquirePoller();
+                try {
+                    level.tick(haveTime);
+                } catch (Throwable throwable) {
+                    return Pair.of(level, throwable);
+                } finally {
+                    level.worldTaskQueueInstance.finalizeScopedTasks();
+                    level.worldTaskQueueInstance.finalizeCyclicalTasks();
+                    level.explosionDensityCache.clear(); // Paper - Optimize explosions // Octopus - Move up
+                    level.worldTaskQueueInstance.releasePoller();
+                }
+                return Pair.of(level, null);
+            };
 
+            me.xiaozhangup.octopus.WorldTickFutureTask task = new me.xiaozhangup.octopus.WorldTickFutureTask(level, tickingFunction);
+            this.worldTickExecutor.submit(task);
+            tasks.add(task);
             profiler.pop();
-            profiler.pop();
-            level.explosionDensityCache.clear(); // Paper - Optimize explosions
+            // Octopus end
+        }
+        // Octopus start
+        me.xiaozhangup.octopus.WorldTickFutureTask tickTask;
+        while((tickTask = tasks.poll()) != null) {
+            try {
+                Pair<ServerLevel, Throwable> result;
+                try {
+                    result = tickTask.get(2, java.util.concurrent.TimeUnit.SECONDS);
+                } catch (java.util.concurrent.TimeoutException te) {
+                    tickTask.createTimeoutReport();
+                    result = tickTask.get();
+                }
+
+                result.getLeft().worldTaskQueueInstance.finalizeCallbackTasks(); // Octopus
+
+                Throwable throwable = result.getRight();
+
+                if (throwable != null) {
+                    CrashReport report = CrashReport.forThrowable(throwable, "Exception ticking world");
+                    result.getLeft().fillReportDetails(report);
+                    throw new ReportedException(report);
+                }
+            } catch (InterruptedException | java.util.concurrent.ExecutionException e) {
+                e.printStackTrace();
+            }
         }
+        // Octopus end
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
         profiler.popPush("connection");
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index c3adbe0753f6a4158f25126ed7443f8a958763d9..6e43269d5c06eb78c429ab8fa76de8c8d09fc5c0 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -116,7 +116,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
         if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkX, chunkZ)) {
             ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.pushChunkWait(this.level, chunkX, chunkZ);
-            this.mainThreadProcessor.managedBlock(completable::isDone);
+            this.level.worldTaskQueueInstance.spinWait(completable::isDone);
             ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.popChunkWait();
         }
 
@@ -353,7 +353,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         CompletableFuture<ChunkResult<ChunkAccess>> serverFuture;
         if (isMainThread) {
             serverFuture = this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate);
-            this.mainThreadProcessor.managedBlock(serverFuture::isDone);
+            this.level.worldTaskQueueInstance.spinWait(serverFuture::isDone); // Octopus
         } else {
             serverFuture = CompletableFuture.<CompletableFuture<ChunkResult<ChunkAccess>>>supplyAsync(
                     () -> this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate), this.mainThreadProcessor
@@ -431,6 +431,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     }
 
     public boolean pollTask() {
+        if (true) { return false; } // Octopus
         return this.mainThreadProcessor.pollTask();
     }
 
@@ -638,7 +639,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
     @VisibleForTesting
     public int getPendingTasksCount() {
-        return this.mainThreadProcessor.getPendingTasksCount();
+        return this.level.worldTaskQueueInstance.taskQueueLength();
     }
 
     public ChunkGenerator getGenerator() {
@@ -669,7 +670,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
     @Override
     public void onLightUpdate(final LightLayer layer, final SectionPos pos) {
-        this.mainThreadProcessor.execute(() -> {
+        this.level.worldTaskQueueInstance.submitTask(() -> { // Octopus
             ChunkHolder chunk = this.getVisibleChunkIfPresent(pos.chunk().pack());
             if (chunk != null && chunk.sectionLightChanged(layer, pos.y())) {
                 this.chunkHoldersToBroadcast.add(chunk);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index ca1725b9ad40f41e241447e14948ee4116c3e4c9..662665b9e9646dfc2bd01dc6aa9659a11f8a7dd9 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -222,6 +222,25 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private final boolean tickTime;
     private final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
 
+    // Octopus start
+    public final me.xiaozhangup.octopus.WorldTaskQueue worldTaskQueueInstance = new me.xiaozhangup.octopus.WorldTaskQueue(this);
+    public void submitCallbackTask(Runnable task) {
+        worldTaskQueueInstance.submitCallbackTask(task);
+    }
+
+    public void submitScopedTask(Runnable task) {
+        worldTaskQueueInstance.submitScopedTask(task);
+    }
+
+    public void submitCyclicalTask(String id, Runnable task) {
+        worldTaskQueueInstance.submitCyclicalTask(id, task);
+    }
+
+    public Runnable removeCyclicalTask(String id) {
+        return worldTaskQueueInstance.removeCyclicalTask(id);
+    }
+    // Octopus end
+
     // CraftBukkit start
     private final ResourceKey<LevelStem> typeKey;
     public final String bukkitName;
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1530810cdb74810e37e3e22ce4fcda4ee7511055..838af56d8d1263f432ffb8e9a0dd75e2c39cf674 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4201,17 +4201,20 @@ public abstract class Entity
         // Paper end - Fix item duplication and teleport issues
         newEntity.restoreFrom(this);
         this.removeAfterChangingDimensions();
-        newEntity.teleportSetPosition(PositionMoveRotation.of(this), PositionMoveRotation.of(transition), transition.relatives());
-        if (this.inWorld) newLevel.addDuringTeleport(newEntity); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
+        Runnable pendingAdd = () -> { // Octopus
+            newEntity.teleportSetPosition(PositionMoveRotation.of(this), PositionMoveRotation.of(transition), transition.relatives());
+            if (this.inWorld) newLevel.addDuringTeleport(newEntity); // CraftBukkit - Don't spawn the new entity if the current entity isn't spawned
 
-        for (Entity newPassenger : newPassengers) {
-            newPassenger.startRiding(newEntity, true, false);
-        }
+            for (Entity newPassenger : newPassengers) {
+                newPassenger.startRiding(newEntity, true, false);
+            }
 
-        newLevel.resetEmptyTime();
-        transition.postTeleportTransition().onTransition(newEntity);
-        this.teleportSpectators(transition, oldLevel);
-        profiler.pop();
+            newLevel.resetEmptyTime();
+            transition.postTeleportTransition().onTransition(newEntity);
+            this.teleportSpectators(transition, oldLevel);
+            profiler.pop();
+        }; // Octopus
+        newLevel.worldTaskQueueInstance.submitTask(pendingAdd); // Octopus
         return newEntity;
     }
 
diff --git a/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java b/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
index ba61762e09ace201173999b3984046f930ccfe57..e7e47ecff01efafa1d4a1d4af32a0d9f53583e7a 100644
--- a/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -46,9 +46,18 @@ public class GoToPotentialJobSite extends Behavior<Villager> {
             ServerLevel serverLevel = level.getServer().getLevel(globalPos.dimension());
             if (serverLevel != null) {
                 PoiManager manager = serverLevel.getPoiManager();
-                if (manager.exists(pos, p -> true)) {
-                    manager.release(pos);
+                // Octopus start - Fix off world POI accessing
+                Runnable scheduledRelease = () -> {
+                    if (manager.exists(pos, p -> true)) {
+                        manager.release(pos);
+                    }
+                };
+                if (!serverLevel.worldTaskQueueInstance.isOwnedByCurrentThread()) {
+                    serverLevel.worldTaskQueueInstance.submitTask(scheduledRelease);
+                } else {
+                    scheduledRelease.run();
                 }
+                // Octopus end
 
                 level.debugSynchronizers().updatePoi(pos);
             }
diff --git a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
index 09cc0f75c0dea7d9ef9caf9b1f14f46e4da84375..1845baef1c7da6f3b211f4dbdbe0c670d98cbc64 100644
--- a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -35,8 +35,18 @@ public class YieldJobSite {
                             return false;
                         }
 
-                        BlockPos poiPos = i.<GlobalPos>get(potentialJob).pos();
-                        Optional<Holder<PoiType>> poiType = level.getPoiManager().getType(poiPos);
+                        // Octopus start - Fix off world POI accessing
+                        GlobalPos globalPos = i.<GlobalPos>get(potentialJob);
+                        net.minecraft.server.level.ServerLevel targetLevel = net.minecraft.server.MinecraftServer.getServer().getLevel(globalPos.dimension());
+                        if (targetLevel == null) {
+                            return true;
+                        }
+                        BlockPos poiPos = globalPos.pos();
+                        if (!targetLevel.worldTaskQueueInstance.isOwnedByCurrentThread()) {
+                            return true;
+                        }
+                        Optional<Holder<PoiType>> poiType = targetLevel.getPoiManager().getType(poiPos);
+                        // Octopus end
                         if (poiType.isEmpty()) {
                             return true;
                         }
@@ -54,8 +64,8 @@ public class YieldJobSite {
                                 potentialJob.erase();
                                 if (nearbyVillager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
                                     BehaviorUtils.setWalkAndLookTargetMemories(nearbyVillager, poiPos, speedModifier, 1);
-                                    nearbyVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), poiPos));
-                                    level.debugSynchronizers().updatePoi(poiPos);
+                                    nearbyVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(targetLevel.dimension(), poiPos));
+                                    targetLevel.debugSynchronizers().updatePoi(poiPos); // Octopus
                                 }
                             });
                         return true;
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index f1e6e876460e62cee24c4ca4adcf047fb96b1909..204e824c57b6deccb5ef6ee05cccb3c47fb660fe 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -403,8 +403,8 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
             capturedBlockEntities = new java.util.ArrayList<>(level.capturedBlockEntities.values());
             level.capturedBlockEntities.clear();
 
-            treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
-            net.minecraft.world.level.block.SaplingBlock.treeType = null;
+            treeType = net.minecraft.world.level.block.SaplingBlock.treeTypeRT.get(); // Octopus
+            net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(null); // Octopus
         }
 
         DataComponentPatch newPatch = this.components.asPatch();
diff --git a/net/minecraft/world/level/block/MushroomBlock.java b/net/minecraft/world/level/block/MushroomBlock.java
index 8d466e5a3e187cb151a7513de3e78d3350b32277..89a7a69478a1ffd2d94664d7a347ace42717d4ba 100644
--- a/net/minecraft/world/level/block/MushroomBlock.java
+++ b/net/minecraft/world/level/block/MushroomBlock.java
@@ -94,7 +94,7 @@ public class MushroomBlock extends VegetationBlock implements BonemealableBlock
         }
 
         level.removeBlock(pos, false);
-        SaplingBlock.treeType = (this == Blocks.BROWN_MUSHROOM) ? org.bukkit.TreeType.BROWN_MUSHROOM : org.bukkit.TreeType.RED_MUSHROOM; // CraftBukkit
+        SaplingBlock.treeTypeRT.set((this == Blocks.BROWN_MUSHROOM) ? org.bukkit.TreeType.BROWN_MUSHROOM : org.bukkit.TreeType.RED_MUSHROOM); // CraftBukkit // Octopus
         if (feature.get().value().place(level, level.getChunkSource().getGenerator(), random, pos)) {
             return true;
         }
diff --git a/net/minecraft/world/level/block/NetherFungusBlock.java b/net/minecraft/world/level/block/NetherFungusBlock.java
index 82ab930bf438e988d637786bdde88e34ea073a2f..2e7c6db4e9b81a803ddb5dda6a3a126954fc0669 100644
--- a/net/minecraft/world/level/block/NetherFungusBlock.java
+++ b/net/minecraft/world/level/block/NetherFungusBlock.java
@@ -84,9 +84,9 @@ public class NetherFungusBlock extends VegetationBlock implements BonemealableBl
             // CraftBukkit start
             .map((value) -> {
                 if (this == Blocks.WARPED_FUNGUS) {
-                    SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+                    SaplingBlock.treeTypeRT.set(org.bukkit.TreeType.WARPED_FUNGUS); // Octopus
                 } else if (this == Blocks.CRIMSON_FUNGUS) {
-                    SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+                    SaplingBlock.treeTypeRT.set(org.bukkit.TreeType.CRIMSON_FUNGUS); // Octopus
                 }
                 return value;
             })
diff --git a/net/minecraft/world/level/block/SaplingBlock.java b/net/minecraft/world/level/block/SaplingBlock.java
index cbf7845ca64f1b716af3e50ac8716373d2ae3755..ebd48179e48fa8c9afa398f31303080975429368 100644
--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -24,7 +24,7 @@ public class SaplingBlock extends VegetationBlock implements BonemealableBlock {
     public static final IntegerProperty STAGE = BlockStateProperties.STAGE;
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 12.0);
     protected final TreeGrower treeGrower;
-    public static org.bukkit.@org.jspecify.annotations.Nullable TreeType treeType; // CraftBukkit
+    public static final ThreadLocal<org.bukkit.TreeType> treeTypeRT = new ThreadLocal<>(); // CraftBukkit // Octopus (from Folia)
 
     @Override
     public MapCodec<? extends SaplingBlock> codec() {
@@ -68,8 +68,8 @@ public class SaplingBlock extends VegetationBlock implements BonemealableBlock {
                     capturedBlockStates = new java.util.ArrayList<>(level.capturedBlockStates.values());
                     level.capturedBlockStates.clear();
 
-                    treeType = SaplingBlock.treeType;
-                    SaplingBlock.treeType = null;
+                    treeType = SaplingBlock.treeTypeRT.get(); // Octopus
+                    SaplingBlock.treeTypeRT.set(null); // Octopus
                 }
                 if (!capturedBlockStates.isEmpty()) {
                     org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level);
diff --git a/net/minecraft/world/level/block/grower/TreeGrower.java b/net/minecraft/world/level/block/grower/TreeGrower.java
index 2a537a329b635d48865e59c03589b89594ef012e..af6f59d0fb5fd51c941b0dd07b2d5573f085efb5 100644
--- a/net/minecraft/world/level/block/grower/TreeGrower.java
+++ b/net/minecraft/world/level/block/grower/TreeGrower.java
@@ -215,55 +215,57 @@ public final class TreeGrower {
 
     // CraftBukkit start
     private void setTreeType(Holder<ConfiguredFeature<?, ?>> feature) {
+        org.bukkit.TreeType treeType; // Octopus
         if (feature.is(TreeFeatures.OAK) || feature.is(TreeFeatures.OAK_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+            treeType = org.bukkit.TreeType.TREE; // Octopus
         } else if (feature.is(TreeFeatures.HUGE_RED_MUSHROOM)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+            treeType = org.bukkit.TreeType.RED_MUSHROOM; // Octopus
         } else if (feature.is(TreeFeatures.HUGE_BROWN_MUSHROOM)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+            treeType = org.bukkit.TreeType.BROWN_MUSHROOM; // Octopus
         } else if (feature.is(TreeFeatures.JUNGLE_TREE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+            treeType = org.bukkit.TreeType.COCOA_TREE; // Octopus
         } else if (feature.is(TreeFeatures.JUNGLE_TREE_NO_VINE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+            treeType = org.bukkit.TreeType.SMALL_JUNGLE; // Octopus
         } else if (feature.is(TreeFeatures.PINE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+            treeType = org.bukkit.TreeType.TALL_REDWOOD; // Octopus
         } else if (feature.is(TreeFeatures.SPRUCE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+            treeType = org.bukkit.TreeType.REDWOOD; // Octopus
         } else if (feature.is(TreeFeatures.ACACIA)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+            treeType = org.bukkit.TreeType.ACACIA; // Octopus
         } else if (feature.is(TreeFeatures.BIRCH) || feature.is(TreeFeatures.BIRCH_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+            treeType = org.bukkit.TreeType.BIRCH; // Octopus
         } else if (feature.is(TreeFeatures.SUPER_BIRCH_BEES_0002)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+            treeType = org.bukkit.TreeType.TALL_BIRCH; // Octopus
         } else if (feature.is(TreeFeatures.SWAMP_OAK)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+            treeType = org.bukkit.TreeType.SWAMP; // Octopus
         } else if (feature.is(TreeFeatures.FANCY_OAK) || feature.is(TreeFeatures.FANCY_OAK_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+            treeType = org.bukkit.TreeType.BIG_TREE; // Octopus
         } else if (feature.is(TreeFeatures.JUNGLE_BUSH)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+            treeType = org.bukkit.TreeType.JUNGLE_BUSH; // Octopus
         } else if (feature.is(TreeFeatures.DARK_OAK)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+            treeType = org.bukkit.TreeType.DARK_OAK; // Octopus
         } else if (feature.is(TreeFeatures.MEGA_SPRUCE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+            treeType = org.bukkit.TreeType.MEGA_REDWOOD; // Octopus
         } else if (feature.is(TreeFeatures.MEGA_PINE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_PINE;
+            treeType = org.bukkit.TreeType.MEGA_PINE; // Octopus
         } else if (feature.is(TreeFeatures.MEGA_JUNGLE_TREE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+            treeType = org.bukkit.TreeType.JUNGLE; // Octopus
         } else if (feature.is(TreeFeatures.AZALEA_TREE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+            treeType = org.bukkit.TreeType.AZALEA; // Octopus
         } else if (feature.is(TreeFeatures.MANGROVE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MANGROVE;
+            treeType = org.bukkit.TreeType.MANGROVE; // Octopus
         } else if (feature.is(TreeFeatures.TALL_MANGROVE)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_MANGROVE;
+            treeType = org.bukkit.TreeType.TALL_MANGROVE; // Octopus
         } else if (feature.is(TreeFeatures.CHERRY) || feature.is(TreeFeatures.CHERRY_BEES_005)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CHERRY;
+            treeType = org.bukkit.TreeType.CHERRY; // Octopus
         } else if (feature.is(TreeFeatures.PALE_OAK) || feature.is(TreeFeatures.PALE_OAK_BONEMEAL)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.PALE_OAK;
+            treeType = org.bukkit.TreeType.PALE_OAK; // Octopus
         } else if (feature.is(TreeFeatures.PALE_OAK_CREAKING)) {
-            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.PALE_OAK_CREAKING;
+            treeType = org.bukkit.TreeType.PALE_OAK_CREAKING; // Octopus
         } else {
             throw new IllegalArgumentException("Unknown tree generator " + feature);
         }
+        net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(treeType); // Octopus
     }
     // CraftBukkit end
 }
