From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:04:00 +0800
Subject: [PATCH] Disable structure blocks place


diff --git a/net/minecraft/world/level/levelgen/structure/StructureStart.java b/net/minecraft/world/level/levelgen/structure/StructureStart.java
index 063d79ec9975dd30602024c4096579ca5b2f1135..7160578129f9db81373efb20cb77f3fdd25a8127 100644
--- a/net/minecraft/world/level/levelgen/structure/StructureStart.java
+++ b/net/minecraft/world/level/levelgen/structure/StructureStart.java
@@ -102,14 +102,14 @@ public final class StructureStart {
                 org.bukkit.craftbukkit.util.TransformerGeneratorAccess transformerAccess = new org.bukkit.craftbukkit.util.TransformerGeneratorAccess();
                 transformerAccess.setDelegate(level);
                 transformerAccess.setStructureTransformer(new org.bukkit.craftbukkit.util.CraftStructureTransformer(this.generationEventCause, level, structureManager, this.structure, box, chunkPos));
-                for (StructurePiece piece : pieces) {
-                    piece.postProcess(transformerAccess, structureManager, generator, random, box, chunkPos, blockPos);
-                }
+//                for (StructurePiece piece : pieces) {
+//                    piece.postProcess(transformerAccess, structureManager, generator, random, box, chunkPos, blockPos);
+//                }
                 transformerAccess.getStructureTransformer().discard();
             }
             // CraftBukkit end
 
-            this.structure.afterPlace(level, structureManager, generator, random, box, chunkPos, this.pieceContainer);
+//            this.structure.afterPlace(level, structureManager, generator, random, box, chunkPos, this.pieceContainer);
         }
     }
 
