From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:04:00 +0800
Subject: [PATCH] Disable structure blocks place


diff --git a/net/minecraft/world/level/levelgen/structure/StructureStart.java b/net/minecraft/world/level/levelgen/structure/StructureStart.java
index ff9a61db79f8488612f77ceb92226d83cf3ba75e..425b3da69a2f2660f01b499eaaede7cbeee5a4ca 100644
--- a/net/minecraft/world/level/levelgen/structure/StructureStart.java
+++ b/net/minecraft/world/level/levelgen/structure/StructureStart.java
@@ -107,14 +107,14 @@ public final class StructureStart {
                 org.bukkit.craftbukkit.util.TransformerLevelAccessor transformerAccessor = new org.bukkit.craftbukkit.util.TransformerLevelAccessor();
                 transformerAccessor.setDelegate(level);
                 transformerAccessor.setStructureTransformer(new org.bukkit.craftbukkit.util.CraftStructureTransformer(this.generationEventCause, level, structureManager, this.structure, chunkBB, chunkPos));
-                for (StructurePiece piece : intersectingPieces) {
-                    piece.postProcess(transformerAccessor, structureManager, generator, random, chunkBB, chunkPos, referencePos);
-                }
+//                for (StructurePiece piece : intersectingPieces) {
+//                    piece.postProcess(transformerAccessor, structureManager, generator, random, chunkBB, chunkPos, referencePos);
+//                }
                 transformerAccessor.getStructureTransformer().discard();
             }
             // CraftBukkit end
 
-            this.structure.afterPlace(level, structureManager, generator, random, chunkBB, chunkPos, this.pieceContainer);
+//            this.structure.afterPlace(level, structureManager, generator, random, chunkBB, chunkPos, this.pieceContainer);
         }
     }
 
