From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sun, 23 Feb 2025 20:04:55 +0800
Subject: [PATCH] Add EndPortalCreateEvent


diff --git a/net/minecraft/world/item/EnderEyeItem.java b/net/minecraft/world/item/EnderEyeItem.java
index 02f2c38b5f9a503097dea44d5c79518b03b63f9a..83bf20114e04237e36d0a166dadff28c0778c74f 100644
--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -58,7 +58,13 @@ public class EnderEyeItem extends Item {
 
                 for (int i = 0; i < 3; i++) {
                     for (int i1 = 0; i1 < 3; i1++) {
-                        level.setBlock(blockPos.offset(i, 0, i1), Blocks.END_PORTAL.defaultBlockState(), 2);
+                        BlockPos pos = blockPos.offset(i, 0, i1);
+                        org.bukkit.block.Block block = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                        org.bukkit.event.block.EndPortalCreateEvent event = new org.bukkit.event.block.EndPortalCreateEvent(block, block);
+                        level.getCraftServer().getPluginManager().callEvent(event);
+                        if (!event.isCancelled()) {
+                            level.setBlock(pos, Blocks.END_PORTAL.defaultBlockState(), 2);
+                        }
                     }
                 }
 
