From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:04:50 +0800
Subject: [PATCH] Sniffer sniffing suspicious blocks


diff --git a/net/minecraft/data/tags/VanillaBlockTagsProvider.java b/net/minecraft/data/tags/VanillaBlockTagsProvider.java
index b9b92d1dfff11ea05039dfd8a5a2fdc84be6f5d3..6cc8867f5983a77aeb15c3af16a5da53a62243b3 100644
--- a/net/minecraft/data/tags/VanillaBlockTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaBlockTagsProvider.java
@@ -424,7 +424,11 @@ public class VanillaBlockTagsProvider extends IntrinsicHolderTagsProvider<Block>
                 Blocks.MOSS_BLOCK,
                 Blocks.MUD,
                 Blocks.MUDDY_MANGROVE_ROOTS,
-                Blocks.FARMLAND
+                Blocks.FARMLAND,
+                // Octopus start - Sniffer sniffing suspicious blocks
+                Blocks.SAND,
+                Blocks.GRAVEL
+                // Octopus end - Sniffer sniffing suspicious blocks
             );
         this.tag(BlockTags.OCCLUDES_VIBRATION_SIGNALS).addTag(BlockTags.WOOL);
         this.tag(BlockTags.SNOW).add(Blocks.SNOW, Blocks.SNOW_BLOCK, Blocks.POWDER_SNOW);
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index a4d60a0dddabd7981d2db28af6b1d88d64d6e806..6dc9440d11a9cbaa78712083cffa05d5fe50a687 100644
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -275,6 +275,21 @@ public class Sniffer extends Animal {
     private void dropSeed() {
         if (this.level() instanceof ServerLevel serverLevel && this.entityData.get(DATA_DROP_SEED_AT_TICK) == this.tickCount) {
             BlockPos headBlock = this.getHeadBlock();
+            // Octopus start - Sniffer sniffing suspicious blocks
+            net.minecraft.world.level.block.Block block = this.level().getBlockState(headBlock.below()).getBlock();
+            org.bukkit.Material material = block.defaultBlockState().getBukkitMaterial();
+            if (material == org.bukkit.Material.SAND || material == org.bukkit.Material.GRAVEL) {
+                net.minecraft.world.level.block.state.BlockState state =
+                        (material == org.bukkit.Material.SAND ?
+                                net.minecraft.world.level.block.Blocks.SUSPICIOUS_SAND :
+                                net.minecraft.world.level.block.Blocks.SUSPICIOUS_GRAVEL).defaultBlockState();
+
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, headBlock.below(), state)) {
+                    this.level().setBlock(headBlock.below(), state, 2);
+                }
+                return;
+            }
+            // Octopus end - Sniffer sniffing suspicious blocks
             this.dropFromGiftLootTable(serverLevel, BuiltInLootTables.SNIFFER_DIGGING, (serverLevel1, itemStack) -> {
                 ItemEntity itemEntity = new ItemEntity(this.level(), headBlock.getX(), headBlock.getY(), headBlock.getZ(), itemStack);
                 itemEntity.setDefaultPickUpDelay();
