From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:05:13 +0800
Subject: [PATCH] Track how much MSPT each world used


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 20d04d720f325ceb594535a39695143e849f8ff9..ade1704fbc3a3f094dc047facbcd6bf4fe3dccbe 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1867,7 +1867,14 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             java.util.concurrent.Callable<Pair<ServerLevel, Throwable>> tickingFunction = () -> {
                 level.worldTaskQueueInstance.acquirePoller();
                 try {
+                    long tickStart = Util.getNanos(); // SparklyPaper - track world's MSPT
                     level.tick(haveTime);
+                    // SparklyPaper start - track world's MSPT
+                    long tickTime = Util.getNanos() - tickStart;
+                    level.tickTimes5s.add(this.tickCount, tickTime);
+                    level.tickTimes10s.add(this.tickCount, tickTime);
+                    level.tickTimes60s.add(this.tickCount, tickTime);
+                    // SparklyPaper end
                 } catch (Throwable throwable) {
                     return Pair.of(level, throwable);
                 } finally {
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 39d006af53ed8144c867d0559cdc5303314d5790..71850f29f2f4753741b629172a31b1ba60e5f57a 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -623,6 +623,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
     // Paper end - chunk tick iteration
 
+        // SparklyPaper start - track world's MSPT
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // SparklyPaper end
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
