From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:05:13 +0800
Subject: [PATCH] Track how much MSPT each world used


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c32579019af386f4c6a8ea61a51300831788c6eb..b5e4e957030ea3d23a4f04716858955a32b6cfd0 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1843,7 +1843,14 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             java.util.concurrent.Callable<Pair<ServerLevel, Throwable>> tickingFunction = () -> {
                 level.worldTaskQueueInstance.acquirePoller();
                 try {
+                    long tickStart = Util.getNanos(); // SparklyPaper - track world's MSPT
                     level.tick(haveTime);
+                    // SparklyPaper start - track world's MSPT
+                    long tickTime = Util.getNanos() - tickStart;
+                    level.tickTimes5s.add(this.tickCount, tickTime);
+                    level.tickTimes10s.add(this.tickCount, tickTime);
+                    level.tickTimes60s.add(this.tickCount, tickTime);
+                    // SparklyPaper end
                 } catch (Throwable throwable) {
                     return Pair.of(level, throwable);
                 } finally {
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 662665b9e9646dfc2bd01dc6aa9659a11f8a7dd9..df20d5c9d7dcaf64588fc21a2bf2630da430179e 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -626,6 +626,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
     // Paper end - chunk tick iteration
 
+        // SparklyPaper start - track world's MSPT
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // SparklyPaper end
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
