From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:05:29 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 71850f29f2f4753741b629172a31b1ba60e5f57a..2000e875759f8af6edcfefb0771e49f7008c22f5 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -629,6 +629,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
     // SparklyPaper end
 
+    // Octopus start - Get the MSPT for a single world
+    public long[] getTickTimes() {
+        return tickTimes5s.getTimes();
+    }
+    // Octopus end - Get the MSPT for a single world
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
