From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:05:29 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index df20d5c9d7dcaf64588fc21a2bf2630da430179e..b0f1b8c5419d5c1179bbaec75366cf7fc49b1e62 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -632,6 +632,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
     // SparklyPaper end
 
+    // Octopus start - Get the MSPT for a single world
+    public long[] getTickTimes() {
+        return tickTimes5s.getTimes();
+    }
+    // Octopus end - Get the MSPT for a single world
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
