From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sun, 23 Feb 2025 20:21:54 +0800
Subject: [PATCH] Track how much MSPT each world used


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 4a0f6fdd8c1dbf3b59cb0e165bec746929f1c804..387ed68d37b1b9aa56bc0c82529af8f0d7b96144 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1744,7 +1744,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             java.util.concurrent.Callable<Pair<ServerLevel, Throwable>> tickingFunction = () -> {
             serverLevel.worldTaskQueueInstance.acquirePoller(); // Octopus
             try {
+                long i = Util.getNanos(); // SparklyPaper - track world's MSPT
                 serverLevel.tick(hasTimeLeft);
+                // SparklyPaper start - track world's MSPT
+                long j = Util.getNanos() - i;
+
+                // These are from the "tickServer" function
+                serverLevel.tickTimes5s.add(this.tickCount, j);
+                serverLevel.tickTimes10s.add(this.tickCount, j);
+                serverLevel.tickTimes60s.add(this.tickCount, j);
+                // SparklyPaper end
             } catch (Throwable var7) {
                 return Pair.of(serverLevel, var7);
             }finally {
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 931664ab426a4d8ca45d718109ddb06c2bd1c925..fe61ff8ac58374c7d5ba2afadb7c515dd19f15d0 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -588,6 +588,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
     // Paper end - chunk tick iteration
 
+        // SparklyPaper start - track world's MSPT
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // SparklyPaper end
+
     public ServerLevel(
         MinecraftServer server,
         Executor dispatcher,
