From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:05:29 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index d8e31ce74631f6c35953088c756de0c19109f6c5..970fdb67766b5f37a3b502351fceeda73e938b39 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -635,6 +635,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
     // SparklyPaper end
 
+    // Octopus start - Get the MSPT for a single world
+    public long[] getTickTimes() {
+        return tickTimes5s.getTimes();
+    }
+    // Octopus end - Get the MSPT for a single world
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
