From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:08:43 +0800
Subject: [PATCH] Add stacktrace logging for chained neighbor updates


diff --git a/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java b/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
index 0a60ff97dc77108997568c75c94327cb6858018f..3297e19019a7f147f9073f91a35ac2616501d047 100644
--- a/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
@@ -63,7 +63,8 @@ public class CollectingNeighborUpdater implements NeighborUpdater {
                 this.stack.push(updates);
             }
         } else if (this.count - 1 == this.maxChainedNeighborUpdates) {
-            LOGGER.error("Too many chained neighbor updates. Skipping the rest. First skipped position: {}", pos.toShortString());
+            LOGGER.error("Too many chained neighbor updates. Skipping the rest. First skipped position: {} ({})", pos.toShortString(), level.getWorld().getName());
+            LOGGER.error("Stack trace: ", new Throwable("Stack trace for chained neighbor updates"));
         }
 
         if (!flag) {
