From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:08:43 +0800
Subject: [PATCH] Add stacktrace logging for chained neighbor updates


diff --git a/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java b/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
index deb62cf4b878f34ba13f83faf711c0666f12e6fa..ff2be44d283974668d3c4c9e87a9ae7289e6a500 100644
--- a/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
@@ -76,7 +76,8 @@ public class CollectingNeighborUpdater implements NeighborUpdater {
                 this.stack.push(update);
             }
         } else if (this.count - 1 == this.maxChainedNeighborUpdates) {
-            LOGGER.error("Too many chained neighbor updates. Skipping the rest. First skipped position: {}", pos.toShortString());
+            LOGGER.error("Too many chained neighbor updates. Skipping the rest. First skipped position: {} ({})", pos.toShortString(), level.getWorld().getName());
+            LOGGER.error("Stack trace: ", new Throwable("Stack trace for chained neighbor updates"));
         }
 
         if (!runningAlready) {
