From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:10:03 +0800
Subject: [PATCH] Redundant move packet optimization


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 60af82925d2f7e605bf7340cf01f1a5668f4d725..939f5cb88d4880e77d1d059dc0ad58d546d01ec6 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -186,18 +186,23 @@ public class ServerEntity {
                     packet = ClientboundEntityPositionSyncPacket.of(this.entity);
                     sentPosition = true;
                     sentRotation = true;
-                } else if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
-                    if (pos) {
-                        packet = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)xa, (short)ya, (short)za, this.entity.onGround());
-                        sentPosition = true;
-                    } else if (shouldSendRotation) {
-                        packet = new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround());
-                        sentRotation = true;
-                    }
+                // Gale start - Airplane - better checking for useless move packets
                 } else {
-                    packet = new ClientboundMoveEntityPacket.PosRot(this.entity.getId(), (short)xa, (short)ya, (short)za, yRotn, xRotn, this.entity.onGround());
-                    sentPosition = true;
-                    sentRotation = true;
+                    if (pos || shouldSendRotation || this.entity instanceof AbstractArrow) {
+                        if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
+                            if (pos) {
+                                packet = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)xa, (short)ya, (short)za, this.entity.onGround());
+                                sentPosition = true;
+                            } else if (shouldSendRotation) {
+                                packet = new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround());
+                                sentRotation = true;
+                            }
+                        } else {
+                            packet = new ClientboundMoveEntityPacket.PosRot(this.entity.getId(), (short)xa, (short)ya, (short)za, yRotn, xRotn, this.entity.onGround());
+                            sentPosition = true;
+                            sentRotation = true;
+                        }
+                    }
                 }
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity livingEntity && livingEntity.isFallFlying()) {
@@ -220,6 +225,7 @@ public class ServerEntity {
                         }
                     }
                 }
+                // Gale end - Airplane - better checking for useless move packets
 
                 if (packet != null) {
                     this.synchronizer.sendToTrackingPlayers(packet);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 838af56d8d1263f432ffb8e9a0dd75e2c39cf674..0981125224f34dcd54be7ef7769d3c11a1631623 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -280,6 +280,7 @@ public abstract class Entity
     public float yRotO;
     public float xRotO;
     private AABB bb = INITIAL_AABB;
+    private boolean boundingBoxChanged = false; // Gale - VMP - skip entity move if movement is zero
     public boolean onGround;
     public boolean horizontalCollision;
     public boolean verticalCollision;
@@ -1136,6 +1137,11 @@ public abstract class Entity
     // Paper end - detailed watchdog information
 
     public void move(final MoverType moverType, Vec3 delta) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.boundingBoxChanged && delta.equals(Vec3.ZERO)) {
+            return;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         final Vec3 originalMovement = delta; // Paper - Expose pre-collision velocity
         // Paper start - detailed watchdog information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
@@ -4583,6 +4589,11 @@ public abstract class Entity
     }
 
     public final void setBoundingBox(final AABB bb) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.bb.equals(bb)) {
+            this.boundingBoxChanged = true;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
                 minY = bb.minY,
