From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Tue, 11 Mar 2025 23:17:59 +0800
Subject: [PATCH] Custom PlayerDataStorage


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index b3578ad892cb9110ccf6e0f29373e1f96e369078..f9ffb0b111e689c3a20b0e2d1b7ccd9831bc074a 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -207,6 +207,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     private Consumer<ProfileResults> onMetricsRecordingStopped = results -> this.stopRecordingMetrics();
     private Consumer<Path> onMetricsRecordingFinished = path -> {};
     private boolean willStartRecordingMetrics;
+    // Octopus start - Custom DataSource
+    @Nullable
+    public me.xiaozhangup.octopus.PlayerJsonDataSource achievementsSource = null;
+    @Nullable
+    public me.xiaozhangup.octopus.PlayerJsonDataSource statsSource = null;
+    // Octopus end - Custom DataSource
     @Nullable
     private MinecraftServer.TimeProfiler debugCommandProfiler;
     private boolean debugCommandProfilerDelayStart;
diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 2a60512bcf922e1356d7de0e4fce337c46d1315b..ac7324542480a9483ef87a72c1994895f720a067 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -111,6 +111,15 @@ public class PlayerAdvancements {
     }
 
     private void load(ServerAdvancementManager manager) {
+        if (playerList.getServer().achievementsSource != null) {
+            String jsonData = playerList.getServer().achievementsSource.load(player.getUUID().toString());
+            if (jsonData != null) {
+                JsonElement jsonElement = com.google.gson.JsonParser.parseString(jsonData);
+                PlayerAdvancements.Data data = this.codec.parse(JsonOps.INSTANCE, jsonElement).getOrThrow(JsonParseException::new);
+                this.applyFrom(manager, data);
+            }
+            return;
+        }
         if (Files.isRegularFile(this.playerSavePath)) {
             try (JsonReader jsonReader = new JsonReader(Files.newBufferedReader(this.playerSavePath, StandardCharsets.UTF_8))) {
                 jsonReader.setLenient(false);
@@ -133,6 +142,10 @@ public class PlayerAdvancements {
         JsonElement jsonElement = this.codec.encodeStart(JsonOps.INSTANCE, this.asData()).getOrThrow();
 
         try {
+            if (playerList.getServer().achievementsSource != null) {
+                playerList.getServer().achievementsSource.save(GSON.toJson(jsonElement), player.getUUID().toString());
+                return;
+            }
             FileUtil.createDirectoriesSafe(this.playerSavePath.getParent());
 
             try (Writer bufferedWriter = Files.newBufferedWriter(this.playerSavePath, StandardCharsets.UTF_8)) {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index c9dbe659374e3ce140316116e05110567e44b810..c01711b5dd8d435da7b9c03d8b8d4686fd2ad8c6 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1310,7 +1310,7 @@ public abstract class PlayerList {
                 }
             }
 
-            serverStatsCounter = new ServerStatsCounter(this.server, file1);
+            serverStatsCounter = new ServerStatsCounter(this.server, file1, uuid.toString()); // Octopus
             // this.stats.put(uuid, serverStatsCounter); // CraftBukkit
         }
 
diff --git a/net/minecraft/stats/ServerStatsCounter.java b/net/minecraft/stats/ServerStatsCounter.java
index 6c5205fe1dc6387a77b1edbdcab748d39e775d7f..9354da75b4a75340ff257254f5621021a9768002 100644
--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -38,11 +38,19 @@ public class ServerStatsCounter extends StatsCounter {
     private final MinecraftServer server;
     private final File file;
     private final Set<Stat<?>> dirty = Sets.newHashSet();
+    private final String uuid;
 
-    public ServerStatsCounter(MinecraftServer server, File file) {
+    public ServerStatsCounter(MinecraftServer server, File file, String uuid) {
         this.server = server;
         this.file = file;
-        if (file.isFile()) {
+        this.uuid = uuid;
+        if (server.statsSource != null) {
+            String json = server.statsSource.load(uuid);
+            if (json != null) {
+                this.parseLocal(server.getFixerUpper(), json);
+            }
+        }
+        else if (file.isFile()) {
             try {
                 this.parseLocal(server.getFixerUpper(), FileUtils.readFileToString(file));
             } catch (IOException var4) {
@@ -67,6 +75,10 @@ public class ServerStatsCounter extends StatsCounter {
     public void save() {
         if (org.spigotmc.SpigotConfig.disableStatSaving) return; // Spigot
         try {
+            if (server.statsSource != null) {
+                server.statsSource.save(this.toJson(), this.uuid);
+                return;
+            }
             FileUtils.writeStringToFile(this.file, this.toJson());
         } catch (IOException var2) {
             LOGGER.error("Couldn't save stats", (Throwable)var2);
diff --git a/net/minecraft/world/level/storage/PlayerDataStorage.java b/net/minecraft/world/level/storage/PlayerDataStorage.java
index c44110b123ba5912af18faf0065e9ded780da9b7..6bd1bcd2e5808f1557747a32702daa755f4048e4 100644
--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -3,6 +3,7 @@ package net.minecraft.world.level.storage;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import java.io.File;
+import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
@@ -25,6 +26,13 @@ public class PlayerDataStorage {
     private final File playerDir;
     protected final DataFixer fixerUpper;
     private static final DateTimeFormatter FORMATTER = FileNameDateFormatter.create();
+    private @org.jetbrains.annotations.Nullable me.xiaozhangup.octopus.PlayerDataSource dataSource = null; // Octopus
+
+    // Octopus start - PlayerDataSource
+    public void setDataSource(@org.jetbrains.annotations.Nullable me.xiaozhangup.octopus.PlayerDataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+    // Octopus end - PlayerDataSource
 
     public PlayerDataStorage(LevelStorageSource.LevelStorageAccess levelStorageAccess, DataFixer fixerUpper) {
         this.fixerUpper = fixerUpper;
@@ -36,6 +44,12 @@ public class PlayerDataStorage {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
         try {
             CompoundTag compoundTag = player.saveWithoutId(new CompoundTag());
+            // Octopus start - PlayerDataSource
+            if (dataSource != null) {
+                dataSource.save(player, compoundTag);
+                return;
+            }
+            // Octopus end - PlayerDataSource
             Path path = this.playerDir.toPath();
             Path path1 = Files.createTempFile(path, player.getStringUUID() + "-", ".dat");
             NbtIo.writeCompressed(compoundTag, path1);
@@ -90,6 +104,21 @@ public class PlayerDataStorage {
     }
 
     public Optional<CompoundTag> load(Player player) {
+        // Octopus start - PlayerDataSource
+        try {
+            if (dataSource != null) {
+                Optional<CompoundTag> data = dataSource.load(player);
+                if (data.isPresent()) {
+                    return data.map(tag -> {
+                        player.load(tag);
+                        return tag;
+                    });
+                }
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Failed to load player data for {}", player.getScoreboardName(), e); // Paper - Print exception
+        }
+        // Octopus end - PlayerDataSource
         // CraftBukkit start
         return this.load(player.getName().getString(), player.getStringUUID()).map((tag) -> {
             if (player instanceof ServerPlayer serverPlayer) {
@@ -107,6 +136,22 @@ public class PlayerDataStorage {
     }
 
     public Optional<CompoundTag> load(String name, String uuid) {
+        // Octopus start - PlayerDataSource
+        try {
+            if (dataSource != null) {
+                Optional<CompoundTag> data = dataSource.load(name, uuid);
+                if (data.isPresent()) {
+                    return data.map(compoundTag -> {
+                        int dataVersion = NbtUtils.getDataVersion(compoundTag, -1);
+                        compoundTag = ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.PLAYER, compoundTag, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().getDataVersion().getVersion());
+                        return compoundTag;
+                    });
+                }
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Failed to load player data for {}", name, e); // Paper - Print exception
+        }
+        // Octopus end - PlayerDataSource
         // CraftBukkit end
         Optional<CompoundTag> optional = this.load(name, uuid, ".dat"); // CraftBukkit
         if (optional.isEmpty()) {
