From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:40:20 +0800
Subject: [PATCH] Prevent experience orbs from referencing offline players


diff --git a/net/minecraft/world/entity/ExperienceOrb.java b/net/minecraft/world/entity/ExperienceOrb.java
index 700c9c0d40dee1df6c97f7c63ad731f16bc5f3e7..6db0f1f35fb0c78b738899a4806ee3c17d3fdd1c 100644
--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -202,7 +202,7 @@ public class ExperienceOrb extends Entity {
 
     private void followNearbyPlayer() {
         Player prevTarget = this.followingPlayer; // CraftBukkit - store old target
-        if (this.followingPlayer == null || this.followingPlayer.isSpectator() || this.followingPlayer.distanceToSqr(this) > 64.0) {
+        if (this.followingPlayer == null || this.followingPlayer.isSpectator() || this.followingPlayer.distanceToSqr(this) > 64.0 || this.followingPlayer.isRemoved()) {
             Player nearestPlayer = this.level().getNearestPlayer(this, 8.0);
             if (nearestPlayer != null && !nearestPlayer.isSpectator() && !nearestPlayer.isDeadOrDying()) {
                 this.followingPlayer = nearestPlayer;
