From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:41:24 +0800
Subject: [PATCH] ShulkerBullet tick baffle


diff --git a/net/minecraft/world/entity/projectile/ShulkerBullet.java b/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 27c143bcc0b64b951027be0a6cac0bba9f641cda..2b014bd5a8c582ac9e5ad6be33bd1cefbe3b3577 100644
--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -35,6 +35,10 @@ import org.jspecify.annotations.Nullable;
 
 public class ShulkerBullet extends Projectile {
     private static final double SPEED = 0.15;
+    // Octopus start - ShulkerBullet tick baffle
+    private static final int BAFFLE_RATE = 3;
+    private int tickCounter = 0;
+    // Octopus end - ShulkerBullet tick baffle
     private @Nullable EntityReference<Entity> finalTarget;
     public @Nullable Direction currentMoveDirection;
     public int flightSteps;
@@ -220,9 +224,13 @@ public class ShulkerBullet extends Projectile {
                     movement.add((this.targetDeltaX - movement.x) * 0.2, (this.targetDeltaY - movement.y) * 0.2, (this.targetDeltaZ - movement.z) * 0.2)
                 );
             }
+        }
 
+        // Octopus start - ShulkerBullet tick baffle
+        if (!this.level().isClientSide() && (tickCounter % BAFFLE_RATE) == 0) {
             hitResult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
         }
+        // Octopus end - ShulkerBullet tick baffle
 
         Vec3 movement = this.getDeltaMovement();
         this.setPos(this.position().add(movement));
@@ -261,6 +269,8 @@ public class ShulkerBullet extends Projectile {
                 }
             }
         }
+
+        tickCounter = (tickCounter + 1) % BAFFLE_RATE; // Octopus - ShulkerBullet tick baffle
     }
 
     @Override
