From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:42:32 +0800
Subject: [PATCH] KioCG Chunk API


diff --git a/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java b/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java
index 4baf46256e90a87d684e57ff623076a8e9395f31..c9c3a42ac456288d966d280f6091bd77cba57c7e 100644
--- a/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java
+++ b/ca/spottedleaf/moonrise/paper/util/BaseChunkSystemHooks.java
@@ -118,6 +118,7 @@ public abstract class BaseChunkSystemHooks implements ca.spottedleaf.moonrise.co
 
     @Override
     public void onChunkNotTicking(final LevelChunk chunk, final ChunkHolder holder) {
+        chunk.getChunkHot().clear(); // KioCG
         ((ChunkSystemServerLevel)((ServerLevel)chunk.getLevel())).moonrise$getTickingChunks().remove(chunk);
     }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c0d6f5afc4dae3e1e9d71a17686dd66c4a968b2c..be98b92e6b8b7d6006c99d49db7c9cb50113e17f 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -25,17 +25,7 @@ import java.nio.file.FileStore;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.security.KeyPair;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 import java.util.Map.Entry;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
@@ -1616,7 +1606,30 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
         this.tickCount++;
         this.tickRateManager.tick();
+        // KioCG start - ChunkHot
+        final List<net.minecraft.world.level.chunk.LevelChunk> chunks = new ArrayList<>();
+        final List<ServerLevel> allLevelsCopy = new ArrayList<>(this.levels.values());
+        final int chunkHotPhase = Math.floorMod(this.tickCount, 20);
+        for (ServerLevel level : allLevelsCopy) {
+            for (net.minecraft.world.level.chunk.LevelChunk chunk : level.moonrise$getTickingChunks()) {
+                if (Math.floorMod(level.dimension().identifier().hashCode() + chunk.getPos().x() + 3 * chunk.getPos().z(), 20) != chunkHotPhase) {
+                    continue;
+                }
+
+                chunk.getChunkHot().nextTick();
+                chunk.getChunkHot().start();
+                chunks.add(chunk);
+            }
+        }
+        //KioCG end
         this.tickChildren(haveTime);
+        // KioCG start - ChunkHot
+        for (net.minecraft.world.level.chunk.LevelChunk chunk : chunks) {
+            if (chunk.getChunkHot().isStarted()) {
+                chunk.getChunkHot().stop();
+            }
+        }
+        //KioCG end
         if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 970fdb67766b5f37a3b502351fceeda73e938b39..270678b309e0b9d3a9debd359cad78a5442750e4 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1439,11 +1439,14 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         profiler.incrementCounter("tickNonPassenger");
         entity.commonTick();
         final boolean isActive = io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2
+            LevelChunk levelChunk = entity.shouldTickHot() ? this.getChunkIfLoaded(entity.moonrise$getSectionX(),entity.moonrise$getSectionZ()) : null; // KioCG
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
         if (isActive) { // Paper - EAR 2
         entity.tick();
         entity.postTick(); // CraftBukkit
         } else {entity.inactiveTick();} // Paper - EAR 2
         profiler.pop();
+        } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
 
         for (Entity passenger : entity.getPassengers()) {
             this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
@@ -1461,6 +1464,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         if (entity.isRemoved() || entity.getVehicle() != vehicle) {
             entity.stopRiding();
         } else if (entity instanceof Player || this.entityTickList.contains(entity)) {
+            LevelChunk levelChunk = !(entity instanceof Player) ? this.getChunkIfLoaded(entity.blockPosition()) : null; // KioCG
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
             ProfilerFiller profiler = Profiler.get();
             profiler.push(entity.typeHolder()::getRegisteredName);
             profiler.incrementCounter("tickPassenger");
@@ -1477,6 +1482,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             }
             // Paper end - EAR 2
             profiler.pop();
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
 
             for (Entity passenger : entity.getPassengers()) {
                 this.tickPassenger(entity, passenger, isActive); // Paper - EAR 2
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 23b3a82ab3e1c6abc85000d9285470cb12300a6b..8b6388cb69d875acfd919952de8dea28cf62a4db 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -776,7 +776,34 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         if (this.postEffectsDirty) {
             this.sendPostEffects();
         }
+
+        // KioCG start - ChunkHot
+        if (Math.floorMod(this.tickCount + this.getUUID().hashCode(), 20) == 0) {
+            this.nearbyChunkHot = this.refreshNearbyChunkHot();
+        }
+        // KioCG end
+    }
+
+    // KioCG start - ChunkHot
+    private volatile long nearbyChunkHot = 0;
+
+    public long getNearbyChunkHot() { return this.nearbyChunkHot; }
+
+    private long refreshNearbyChunkHot() {
+        long total = 0L;
+        int searchRadius = ((ServerLevel) this.level()).moonrise$getViewDistanceHolder().getViewDistances().tickViewDistance();
+        for (int i = this.moonrise$getSectionX() - searchRadius; i <= this.moonrise$getSectionX() + searchRadius; ++i) {
+            for (int j = this.moonrise$getSectionZ() - searchRadius; j <= this.moonrise$getSectionZ() + searchRadius; ++j) {
+                net.minecraft.world.level.chunk.LevelChunk targetChunk = this.level().getChunkIfLoaded(i, j);
+                if (targetChunk != null) {
+                    total += targetChunk.getChunkHot().getAverage();
+                }
+            }
+        }
+        return total;
     }
+    // KioCG end
+
 
     private void updatePlayerAttributes() {
         AttributeInstance blockInteractionRange = this.getAttribute(Attributes.BLOCK_INTERACTION_RANGE);
diff --git a/net/minecraft/world/entity/AreaEffectCloud.java b/net/minecraft/world/entity/AreaEffectCloud.java
index c5a212e73c2efeb1f440778cb3f1fa6ee8165e63..52f0b0ad01d852f5bd17202eeaccd9a538cd5085 100644
--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -415,4 +415,11 @@ public class AreaEffectCloud extends Entity implements TraceableEntity {
             return super.applyImplicitComponent(type, value);
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 72aec91441106bca870bbea2556e4ab17539eee0..80f2c1718b633937dab64841c4639e67670ee081 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5638,4 +5638,6 @@ public abstract class Entity
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end
+
+    public boolean shouldTickHot() { return this.tickCount > 20 * 10 && this.isAlive(); } // KioCG
 }
diff --git a/net/minecraft/world/entity/LightningBolt.java b/net/minecraft/world/entity/LightningBolt.java
index c756d28b396520264336832f17527981248556f3..90630d3528cb6c07b5242fe868a6b610df241570 100644
--- a/net/minecraft/world/entity/LightningBolt.java
+++ b/net/minecraft/world/entity/LightningBolt.java
@@ -285,4 +285,11 @@ public class LightningBolt extends Entity {
     public final boolean hurtServer(final ServerLevel level, final DamageSource source, final float damage) {
         return false;
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 078054c499e7b29d64543f92c1808b5623e3c8e5..53099ce01fef8661580e8157807606c59e8b388d 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1718,4 +1718,11 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     public void swing(final InteractionHand hand, final SwingAnimation animation) {
         this.swing(hand, animation, false);
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && (!this.removeWhenFarAway(0.0) || this.isPersistenceRequired() || this.requiresCustomPersistence());
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/animal/equine/TraderLlama.java b/net/minecraft/world/entity/animal/equine/TraderLlama.java
index 7c92c9335997cee76ff360c51f704ee41a180796..fb4cab83fd23c5561326a58683558e9a4219b0e9 100644
--- a/net/minecraft/world/entity/animal/equine/TraderLlama.java
+++ b/net/minecraft/world/entity/animal/equine/TraderLlama.java
@@ -164,4 +164,11 @@ public class TraderLlama extends Llama {
             super.start();
         }
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return super.shouldTickHot() && !this.canDespawn();
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 4068dc1cea4edca196dc0c58b4804e94dd07881e..2db02b1202297691e27052360698863e0106642b 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1435,6 +1435,13 @@ public abstract class Player extends Avatar implements ContainerUser {
         return true;
     }
 
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
+
     public Either<Player.BedSleepingProblem, Unit> startSleepInBed(
         final AbstractBedBlock bedBlock, final BlockState bedBlockState, final BedRule rule, final BlockPos pos
     ) {
diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index 089dd82e6806f4f1fbff554058dc44f15593e37f..20844e011c471c379824e7b821a6fb74673dec8f 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -567,4 +567,11 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     public interface ProjectileFactory<T extends Projectile> {
         T create(final ServerLevel level, LivingEntity entity, ItemStack itemStack);
     }
+
+    // KioCG start
+    @Override
+    public boolean shouldTickHot() {
+        return false;
+    }
+    // KioCG end
 }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index af4106497a69f9b311970197016c6157475a7d85..519f7415fc6663af9eae3fb1568ed7b3a9b7d63c 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -104,6 +104,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
     // Paper start - rewrite chunk system
     private boolean postProcessingDone;
     private ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkAndHolder;
+    private final com.kiocg.ChunkHot chunkHot = new com.kiocg.ChunkHot(); public com.kiocg.ChunkHot getChunkHot() { return this.chunkHot; } // KioCG
 
     @Override
     public final boolean moonrise$isPostProcessingDone() {
@@ -943,6 +944,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                     try {
                         ProfilerFiller profiler = Profiler.get();
                         profiler.push(this::getType);
+                        LevelChunk.this.chunkHot.startTicking(); // KioCG
                         BlockState blockState = LevelChunk.this.getBlockState(pos);
                         if (this.blockEntity.getType().isValid(blockState)) {
                             this.ticker.tick(LevelChunk.this.level, this.blockEntity.getBlockPos(), blockState, this.blockEntity);
@@ -971,6 +973,8 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                         LevelChunk.this.level.getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, t))); // Paper - ServerExceptionEvent
                         LevelChunk.this.removeBlockEntity(this.getPos());
                         // Paper end - Prevent block entity and entity crashes
+                    }finally {
+                        LevelChunk.this.chunkHot.stopTickingAndCount();
                     }
                 }
             }
diff --git a/net/minecraft/world/level/redstone/NeighborUpdater.java b/net/minecraft/world/level/redstone/NeighborUpdater.java
index 041ed95948ff6bb4cf3610ce4522a6ba105b6812..44dbe6f83c8593c213132feabc7fe27615cb0049 100644
--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -81,7 +81,10 @@ public interface NeighborUpdater {
                 return;
             }
             // CraftBukkit end
+            net.minecraft.world.level.chunk.LevelChunk levelChunk = level.getChunkIfLoaded(pos); // KioCG
+            if (levelChunk != null) levelChunk.getChunkHot().startTicking(); try { // KioCG
             state.handleNeighborChanged(level, pos, changedBlock, orientation, movedByPiston);
+            } finally { if (levelChunk != null) levelChunk.getChunkHot().stopTickingAndCount(); } // KioCG
             // Spigot start
         } catch (StackOverflowError ex) {
             level.lastPhysicsProblem = pos.immutable();
