From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:48:17 +0800
Subject: [PATCH] Add EntityPortalPreparedEvent


diff --git a/net/minecraft/world/level/block/EndPortalBlock.java b/net/minecraft/world/level/block/EndPortalBlock.java
index 6a656113e5206552dbca5de3fa24f788dddbb8e7..860845a6333ab6b8a5fb74c25ec5167ddf74766b 100644
--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -83,6 +83,13 @@ public class EndPortalBlock extends BaseEntityBlock implements Portal {
 
     @Override
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
+        // Octopus start - Add EntityPortalPreparedEvent
+        org.bukkit.event.entity.EntityPortalPreparedEvent portalPreparedEvent = new org.bukkit.event.entity.EntityPortalPreparedEvent(entity.getBukkitEntity(), org.bukkit.PortalType.ENDER);
+        if (!portalPreparedEvent.callEvent()) {
+            entity.portalProcess = null;
+            return null;
+        }
+        // Octopus end - Add EntityPortalPreparedEvent
         LevelData.RespawnData respawnData = currentLevel.getRespawnData();
         boolean fromEnd = currentLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END; // CraftBukkit - SPIGOT-6152: send back to main overworld in custom ends
         ResourceKey<Level> newDimension = fromEnd ? respawnData.dimension() : Level.END;
diff --git a/net/minecraft/world/level/block/NetherPortalBlock.java b/net/minecraft/world/level/block/NetherPortalBlock.java
index 14cca8dae351481328c689d4227d36c00f279b42..016453076f7bfbcbdcfb989bec31b15cddcd5799 100644
--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -143,6 +143,13 @@ public class NetherPortalBlock extends Block implements Portal {
 
     @Override
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
+        // Octopus start - Add EntityPortalPreparedEvent
+        org.bukkit.event.entity.EntityPortalPreparedEvent portalPreparedEvent = new org.bukkit.event.entity.EntityPortalPreparedEvent(entity.getBukkitEntity(), org.bukkit.PortalType.NETHER);
+        if (!portalPreparedEvent.callEvent()) {
+            entity.portalProcess = null;
+            return null;
+        }
+        // Octopus end - Add EntityPortalPreparedEvent
         ResourceKey<Level> newDimension = currentLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? Level.OVERWORLD : Level.NETHER; // CraftBukkit
         ServerLevel newLevel = currentLevel.getServer().getLevel(newDimension);
         // Paper start - Add EntityPortalReadyEvent
