From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Mon, 2 Feb 2026 01:08:26 +0800
Subject: [PATCH] Add PlayerMessagePreSendEvent


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index e9a3512fb96e7000b8f6b5d6e3f73b157da91d76..1d424d62e91dd8ba2234aebba8646310583dbf37 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2430,14 +2430,19 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void sendSystemMessage(final Component message, final boolean overlay) {
         if (this.acceptsSystemMessages(overlay)) {
+            // Octopus start - Add PlayerMessagePreSendEvent
+            org.bukkit.event.player.PlayerMessagePreSendEvent event = new org.bukkit.event.player.PlayerMessagePreSendEvent(ServerPlayer.this.getBukkitEntity(), io.papermc.paper.adventure.PaperAdventure.asAdventure(message));
+            if (!event.callEvent()) return;
+            Component compo = io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getComponent());
+            // Octopus end - Add PlayerMessagePreSendEvent
             this.connection
                 .send(
-                    new ClientboundSystemChatPacket(message, overlay),
+                    new ClientboundSystemChatPacket(compo, overlay), // Octopus - Add PlayerMessagePreSendEvent
                     PacketSendListener.exceptionallySend(
                         () -> {
                             if (this.acceptsSystemMessages(false)) {
                                 int truncatedMessageSize = 256;
-                                String contents = message.getString(256);
+                                String contents = compo.getString(truncatedMessageSize); // Octopus - Add PlayerMessagePreSendEvent
                                 Component contentsMsg = Component.literal(contents).withStyle(ChatFormatting.YELLOW);
                                 return new ClientboundSystemChatPacket(
                                     Component.translatable("multiplayer.message_not_delivered", contentsMsg).withStyle(ChatFormatting.RED), false
