From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Tue, 9 Sep 2025 12:46:48 +0800
Subject: [PATCH] Prevent experience orbs from referencing offline players


diff --git a/net/minecraft/world/entity/ExperienceOrb.java b/net/minecraft/world/entity/ExperienceOrb.java
index a62edee768c30c99213baa6bd736d67ae52b558c..167ccb16552e2e88c70c980f101d2d6c214b1892 100644
--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -200,7 +200,7 @@ public class ExperienceOrb extends Entity {
 
     private void followNearbyPlayer() {
         Player prevTarget = this.followingPlayer; // CraftBukkit - store old target
-        if (this.followingPlayer == null || this.followingPlayer.isSpectator() || this.followingPlayer.distanceToSqr(this) > 64.0) {
+        if (this.followingPlayer == null || this.followingPlayer.isSpectator() || this.followingPlayer.distanceToSqr(this) > 64.0 || this.followingPlayer.isRemoved()) {
             Player nearestPlayer = this.level().getNearestPlayer(this, 8.0);
             if (nearestPlayer != null && !nearestPlayer.isSpectator() && !nearestPlayer.isDeadOrDying()) {
                 this.followingPlayer = nearestPlayer;
