From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Octopus Builder <octopus-builder@example.invalid>
Date: Tue, 2 Jun 2026 22:44:04 +0800
Subject: [PATCH] Limit dropped items per location


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 270678b309e0b9d3a9debd359cad78a5442750e4..3ea5f179e9f5dd98f1c0c9b59034543734692bc5 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -947,6 +947,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                         }
                     }
                 );
+            this.trackedItemEntities.clear(); // Octopus - Limit dropped items per location
             if (this.paperConfig().unsupportedSettings.ticking.blockEntities) { // Paper - option to disable ticking
             profiler.popPush("blockEntities");
             this.tickBlockEntities();
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 8b10b2121b61892eb4f98533c7df1ba549cdc64f..b77b32f5650a7c7611bb53a18cb8e596b7e856e7 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -143,6 +143,18 @@ public class ItemEntity extends Entity implements TraceableEntity {
 
     @Override
     public void tick() {
+        // Octopus start - Limit dropped items per location
+        int counter = this.level().trackedItemEntities.get(this.blockPosition().asLong());
+        if (counter > 72 && this.getDeltaMovement().lengthSqr() < 1.0E-6) {
+            if ((this.totalEntityAge & 1) == 0) {
+                this.level().trackedItemEntities.addTo(this.blockPosition().asLong(), 1);
+                if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
+                    this.pickupDelay--;
+                }
+                return;
+            }
+        }
+        // Octopus end - Limit dropped items per location
         if (this.getItem().isEmpty()) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
@@ -196,6 +208,12 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 }
             }
 
+            // Octopus start - Limit dropped items per location
+            if ((counter > 864 || (counter > 256 && this.totalEntityAge > 20 * 30)) && this.getDeltaMovement().lengthSqr() < 1.0E-6) {
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                return;
+            }
+            // Octopus end - Limit dropped items per location
             boolean moved = Mth.floor(this.xo) != Mth.floor(this.getX())
                 || Mth.floor(this.yo) != Mth.floor(this.getY())
                 || Mth.floor(this.zo) != Mth.floor(this.getZ());
@@ -225,6 +243,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 // CraftBukkit end
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
+            this.level().trackedItemEntities.addTo(this.blockPosition().asLong(), 1); // Octopus - Limit dropped items per location
         }
     }
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 21878bd770f6c0bc10dd0a67be5bd94d7e16f741..7fa823182c4e5de9b4e79cd468c1e28031ada07e 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -155,6 +155,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     @Nullable
     public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
     public final it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<SpawnCategory> ticksPerSpawnCategory = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>();
+    public final it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap trackedItemEntities = new it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap(); // Octopus - Limit dropped items per location
     // Paper start - EAR 2
     public int wakeupInactiveRemainingAnimals;
     public int wakeupInactiveRemainingFlying;
