From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 23:08:07 +0800
Subject: [PATCH] Replace entity fluid height map


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a0eb89f8f8789e40cdef9f9961943a0fe5f73ccc..5d6046235c42b85b7c9480268f10ef80fa6a5bfe 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -288,7 +288,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public int tickCount;
     private int remainingFireTicks;
     public boolean wasTouchingWater;
-    protected Object2DoubleMap<TagKey<Fluid>> fluidHeight = new Object2DoubleArrayMap<>(2);
+    protected Object2DoubleMap<TagKey<Fluid>> fluidHeight = new org.dreeam.leaf.util.map.FluidHeightMap(); // Leaf - Replace fluid height map
     protected boolean wasEyeInWater;
     private final Set<TagKey<Fluid>> fluidOnEyes = new HashSet<>();
     public int invulnerableTime;
