From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Mon, 6 Jul 2026 08:38:38 +0800
Subject: [PATCH] Keep extra data when migrating worlds


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index e2939cc39a8c103e4c4d7b01d39b5546486e4fbc..b79369669495c3fa0a31e30763bde014aaa21d90 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -110,6 +110,19 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            // Octopus start - Keep extra data when migrating worlds
+            if (options.has("convertWorldsOnly")) {
+                Bootstrap.bootStrap();
+                Bootstrap.validate();
+                io.papermc.paper.world.migration.WorldFolderMigration.didInitialLoad = true;
+                io.papermc.paper.world.migration.WorldFolderMigration.convertWorldContainerInPlace(
+                    ((File) options.valueOf("convertWorldsOnly")).toPath(),
+                    RegistryAccess.fromRegistryOfRegistries(net.minecraft.core.registries.BuiltInRegistries.REGISTRY)
+                );
+                return;
+            }
+            // Octopus end - Keep extra data when migrating worlds
+
             io.papermc.paper.plugin.PluginInitializerManager.load(options); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
