From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Sun, 28 Sep 2025 00:03:58 +0800
Subject: [PATCH] Add cause for EntityTeleportEvent


diff --git a/net/minecraft/server/commands/TeleportCommand.java b/net/minecraft/server/commands/TeleportCommand.java
index 6d8ac5f80d5267a2159a464ec331e9b3c65313e2..17ca06e21debbc3528522cefe5b7d6b0f20d19ad 100644
--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -296,7 +296,7 @@ public class TeleportCommand {
                 result = player.teleportTo(level, d, d1, d2, relatives, f2, f3, true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND);
             } else {
                 org.bukkit.Location to = new org.bukkit.Location(level.getWorld(), d, d1, d2, f2, f3);
-                org.bukkit.event.entity.EntityTeleportEvent event = new org.bukkit.event.entity.EntityTeleportEvent(target.getBukkitEntity(), target.getBukkitEntity().getLocation(), to);
+                org.bukkit.event.entity.EntityTeleportEvent event = new org.bukkit.event.entity.EntityTeleportEvent(target.getBukkitEntity(), target.getBukkitEntity().getLocation(), to, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND); // Octopus - add cause
                 level.getCraftServer().getPluginManager().callEvent(event);
                 if (event.isCancelled() || event.getTo() == null) { // Paper
                     return;
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 0268e02d2ef2cb3d699644a804e23a6da4521f4c..ec78130791658027c0f0304c576b7018912737e5 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4296,7 +4296,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
                 if (flag) {
                     if (!(this instanceof ServerPlayer)) {
-                        org.bukkit.event.entity.EntityTeleportEvent teleport = new org.bukkit.event.entity.EntityTeleportEvent(this.getBukkitEntity(), new Location(this.level().getWorld(), x1, y1, z1), new Location(this.level().getWorld(), x, d, z));
+                        org.bukkit.event.entity.EntityTeleportEvent teleport = new org.bukkit.event.entity.EntityTeleportEvent(this.getBukkitEntity(), new Location(this.level().getWorld(), x1, y1, z1), new Location(this.level().getWorld(), x, d, z), cause); // Octopus - add cause
                         this.level().getCraftServer().getPluginManager().callEvent(teleport);
                         if (!teleport.isCancelled() && teleport.getTo() != null) { // Paper
                             Location to = teleport.getTo();
