From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Tue, 7 Jul 2026 21:09:50 +0800
Subject: [PATCH] Fix incorrect level info send


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index db7a7e22d4fa3f06b4674fc3fba8d0387dd278d8..2e8742ccd3705e3b5d4b1e5dade942c509158305 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -291,7 +291,7 @@ public abstract class PlayerList {
 
         //player.refreshEntityData(player); // CraftBukkit - BungeeCord#2321, send complete data to self on spawn // Paper - THIS IS NOT NEEDED ANYMORE
 
-        this.sendLevelInfo(player, level);
+        this.sendLevelInfo(player, player.level()); // Octopus - Fix incorrect level info send
 
         // CraftBukkit start - Only add if the player wasn't moved in the event
         if (player.level() == level && !level.players().contains(player)) {
