From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 27 May 2026 20:16:47 +0800
Subject: [PATCH] Limit dropped items per location


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 208cbe29963eb2963dcd43be160ab2cdcc92cbe8..2781b30759b33831d7b0a34cf32b15f1608274c9 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -888,6 +888,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                         }
                     }
                 );
+            this.trackedItemEntities.clear(); // Octopus - Limit dropped items per location
             profilerFiller.popPush("blockEntities");
             this.tickBlockEntities();
             profilerFiller.pop();
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index e82bb8f2d2b0e771dc371fd7e709116e1f5f204f..61037a587810ee402976305d0408d25e5661a0dc 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -142,6 +142,18 @@ public class ItemEntity extends Entity implements TraceableEntity {
 
     @Override
     public void tick() {
+        // Octopus start - Limit dropped items per location
+        var counter = this.level().trackedItemEntities.get(this.blockPosition().asLong());
+        if (counter > 72 && this.getDeltaMovement().lengthSqr() < 1.0E-6) {
+            if ((this.totalEntityAge & 1) == 0) {
+                this.level().trackedItemEntities.addTo(this.blockPosition().asLong(), 1);
+                if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
+                    this.pickupDelay--;
+                }
+                return;
+            }
+        }
+        // Octopus end - Limit dropped items per location
         if (this.getItem().isEmpty()) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
@@ -192,6 +204,12 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 }
             }
 
+            // Octopus start - Limit dropped items per location
+            if ((counter > 864 || (counter > 256 && this.totalEntityAge > 20 * 30)) && this.getDeltaMovement().lengthSqr() < 1.0E-6) {
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                return;
+            }
+            // Octopus end - Limit dropped items per location
             boolean flag = Mth.floor(this.xo) != Mth.floor(this.getX())
                 || Mth.floor(this.yo) != Mth.floor(this.getY())
                 || Mth.floor(this.zo) != Mth.floor(this.getZ());
@@ -221,6 +239,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 // CraftBukkit end
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
+            this.level().trackedItemEntities.addTo(this.blockPosition().asLong(), 1); // Octopus - Limit dropped items per location
         }
     }
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 107be743e3d1ddf8e965798027ab339849b37b9a..c89ae86be09178da04655b81635ea97c9b7aeed6 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -151,6 +151,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     @Nullable
     public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
     public final it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<SpawnCategory> ticksPerSpawnCategory = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>();
+    public final it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap trackedItemEntities = new it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap(); // Octopus - Limit dropped items per location
     // Paper start - EAR 2
     public int wakeupInactiveRemainingAnimals;
     public int wakeupInactiveRemainingFlying;
