From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Tue, 9 Sep 2025 12:28:50 +0800
Subject: [PATCH] Apply a fixed-size collection to CombatTracker


diff --git a/net/minecraft/world/damagesource/CombatTracker.java b/net/minecraft/world/damagesource/CombatTracker.java
index 4cec197634fac341cca1ed108f1ecb0561f72461..72d29fc8ca40a64c3fdb1baf9be484b7b5480430 100644
--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -22,7 +22,7 @@ public class CombatTracker {
     private static final Style INTENTIONAL_GAME_DESIGN_STYLE = Style.EMPTY
         .withClickEvent(new ClickEvent.OpenUrl(CommonLinks.INTENTIONAL_GAME_DESIGN_BUG))
         .withHoverEvent(new HoverEvent.ShowText(Component.literal("MCPE-28723")));
-    public final List<CombatEntry> entries = Lists.newArrayList();
+    public final List<CombatEntry> entries = new me.xiaozhangup.octopus.util.FixedSizeList<>(24); // Octopus - Apply a fixed-size collection to CombatTracker (MC-298464)
     public final LivingEntity mob;
     private int lastDamageTime;
     private int combatStartTime;
