From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Wed, 31 Dec 2025 18:53:46 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index faa57d7a4dc75a4c8d5c3feba7ae0b1b1fc72547..97f13d9b1d690b3ed8e0c6c65c728492374d96e0 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2044,4 +2044,24 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    @Override
+    public void submitCallbackTask(Runnable task) {
+        world.submitCallbackTask(task);
+    }
+
+    @Override
+    public void submitScopedTask(Runnable task) {
+        world.submitScopedTask(task);
+    }
+
+    @Override
+    public void submitCyclicalTask(String id, Runnable task) {
+        world.submitCyclicalTask(id, task);
+    }
+
+    @Override
+    public Runnable removeCyclicalTask(String id) {
+        return world.removeCyclicalTask(id);
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index bf98fc87b0a2146455f7c56dca6584bb6ae6fb76..8bbbb6429ab482f892bfab3fb27483909afde22f 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -518,8 +518,8 @@ public class CraftBlock implements Block {
             capturedBlockStates = new ArrayList<>(world.capturedBlockStates.values());
             world.capturedBlockStates.clear();
 
-            treeType = SaplingBlock.treeType;
-            SaplingBlock.treeType = null;
+            treeType = SaplingBlock.treeTypeRT.get();
+            SaplingBlock.treeTypeRT.set(null);
         }
 
         if (!capturedBlockStates.isEmpty()) {
