From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Thu, 24 Jul 2025 18:10:12 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index aa0a61d7f3128543b698647663fa4a9af2a8255a..8704d9af920fe3f1e325841baa29e6cbe4ba6e4f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2375,6 +2375,26 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return (this.getHandle().getDragonFight() == null) ? null : new CraftDragonBattle(this.getHandle().getDragonFight());
     }
 
+    @Override
+    public void submitCallbackTask(Runnable task) {
+        world.submitCallbackTask(task);
+    }
+
+    @Override
+    public void submitScopedTask(Runnable task) {
+        world.submitScopedTask(task);
+    }
+
+    @Override
+    public void submitCyclicalTask(String id, Runnable task) {
+        world.submitCyclicalTask(id, task);
+    }
+
+    @Override
+    public Runnable removeCyclicalTask(String id) {
+        return world.removeCyclicalTask(id);
+     }
+
     @Override
     public Collection<GeneratedStructure> getStructures(int x, int z) {
         return this.getStructures(x, z, struct -> true);
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 6809f9dcf8510c714145d99d250eb69f98d9bf27..7292ef3b2adf7f0364a4e8b062c931d2de7fb0fd 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -559,8 +559,8 @@ public class CraftBlock implements Block {
         world.captureTreeGeneration = false;
 
         if (!world.capturedBlockStates.isEmpty()) {
-            TreeType treeType = SaplingBlock.treeType;
-            SaplingBlock.treeType = null;
+            TreeType treeType = SaplingBlock.treeTypeRT.get();
+            SaplingBlock.treeTypeRT.set(null);
             List<BlockState> states = new ArrayList<>(world.capturedBlockStates.values());
             world.capturedBlockStates.clear();
             StructureGrowEvent structureEvent = null;
