From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Wed, 31 Dec 2025 18:53:46 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index a26249fbae41b8915ab8b3e876485213ab992599..f3a5d2a57c77c48c9d7a3d24734c5ac0a0ad4e5e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2049,4 +2049,24 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    @Override
+    public void submitCallbackTask(Runnable task) {
+        world.submitCallbackTask(task);
+    }
+
+    @Override
+    public void submitScopedTask(Runnable task) {
+        world.submitScopedTask(task);
+    }
+
+    @Override
+    public void submitCyclicalTask(String id, Runnable task) {
+        world.submitCyclicalTask(id, task);
+    }
+
+    @Override
+    public Runnable removeCyclicalTask(String id) {
+        return world.removeCyclicalTask(id);
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 0ab39922e53ebeb044958f99b7f0f06391d78d31..3951468107ef897f454d2b082fcde1a3a79f0e70 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -532,8 +532,8 @@ public class CraftBlock implements Block {
             capturedBlockStates = new ArrayList<>(world.capturedBlockStates.values());
             world.capturedBlockStates.clear();
 
-            treeType = SaplingBlock.treeType;
-            SaplingBlock.treeType = null;
+            treeType = SaplingBlock.treeTypeRT.get();
+            SaplingBlock.treeTypeRT.set(null);
         }
 
         if (!capturedBlockStates.isEmpty()) {
