From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:56:12 +0800
Subject: [PATCH] Add clientConnected API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index f11ad0f2cdbb128dfa1f572903ad1b16b80c3e87..7fdabb52223bd4b22dbb130642fc205d40f2ab98 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -247,6 +247,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     private int hash = 0;
     private double health = 20;
     private boolean scaledHealth = false;
+    private boolean clientConnected = true;
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
@@ -2781,6 +2782,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         return this.getHandle().allowsListing();
     }
 
+    // Octopus start
+    @Override
+    public boolean clientConnected() {
+        return clientConnected;
+    }
+
+    public void setClientConnected(boolean clientConnected) {
+        this.clientConnected = clientConnected;
+    }
+    // Octopus end
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
