From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:56:12 +0800
Subject: [PATCH] Add clientConnected API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 73f66b893c2d9541e78118150279be305f47a663..5574f13ccfe36651223ebcb3831eda0b605e395d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -243,6 +243,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     private int hash = 0;
     private double health = 20;
     private boolean scaledHealth = false;
+    private boolean clientConnected = true;
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
@@ -2756,6 +2757,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         return this.getHandle().allowsListing();
     }
 
+    // Octopus start
+    @Override
+    public boolean clientConnected() {
+        return clientConnected;
+    }
+
+    public void setClientConnected(boolean clientConnected) {
+        this.clientConnected = clientConnected;
+    }
+    // Octopus end
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
