From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:56:12 +0800
Subject: [PATCH] Add clientConnected API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ad51146ef274106d542bf416cd47d0ebcd17e7c6..76c6860245defb657256aa652ae170ef99a58cec 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -241,6 +241,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     private int hash = 0;
     private double health = 20;
     private boolean scaledHealth = false;
+    private boolean clientConnected = true;
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
@@ -2754,6 +2755,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         return this.getHandle().allowsListing();
     }
 
+    // Octopus start
+    @Override
+    public boolean clientConnected() {
+        return clientConnected;
+    }
+
+    public void setClientConnected(boolean clientConnected) {
+        this.clientConnected = clientConnected;
+    }
+    // Octopus end
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
