From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Wed, 31 Dec 2025 22:56:41 +0800
Subject: [PATCH] Track how much MSPT each world used


diff --git a/src/main/java/io/papermc/paper/command/PaperMSPTCommand.java b/src/main/java/io/papermc/paper/command/PaperMSPTCommand.java
index c4e9a5771f8015c75bbd16c04040607d8c7e5fbd..252245a217bd792a44026b3378086d760d9e4588 100644
--- a/src/main/java/io/papermc/paper/command/PaperMSPTCommand.java
+++ b/src/main/java/io/papermc/paper/command/PaperMSPTCommand.java
@@ -76,6 +76,45 @@ public class PaperMSPTCommand {
                 )
             )
         );
+        // SparklyPaper start - track world's MSPT
+        sender.sendMessage(text());
+        sender.sendMessage(text().content("World tick times ").color(GOLD)
+            .append(text().color(YELLOW)
+                .append(
+                    text("("),
+                    text("avg", GRAY),
+                    text("/"),
+                    text("min", GRAY),
+                    text("/"),
+                    text("max", GRAY),
+                    text(")")
+                )
+            ).append(
+                text(" from last 5s"),
+                text(",", GRAY),
+                text(" 10s"),
+                text(",", GRAY),
+                text(" 1m"),
+                text(":", YELLOW)
+            )
+        );
+        for (net.minecraft.server.level.ServerLevel serverLevel : server.getAllLevels()) {
+            List<Component> worldTimes = new ArrayList<>();
+            worldTimes.addAll(eval(serverLevel.tickTimes5s.getTimes()));
+            worldTimes.addAll(eval(serverLevel.tickTimes10s.getTimes()));
+            worldTimes.addAll(eval(serverLevel.tickTimes60s.getTimes()));
+
+            sender.sendMessage(text().content("◴ " + serverLevel.getWorld().getName() + ": ").color(GOLD)
+                .append(text().color(GRAY)
+                    .append(
+                        worldTimes.get(0), SLASH, worldTimes.get(1), SLASH, worldTimes.get(2), text(", ", YELLOW),
+                        worldTimes.get(3), SLASH, worldTimes.get(4), SLASH, worldTimes.get(5), text(", ", YELLOW),
+                        worldTimes.get(6), SLASH, worldTimes.get(7), SLASH, worldTimes.get(8)
+                    )
+                )
+            );
+        }
+        // SparklyPaper end
         return Command.SINGLE_SUCCESS;
     }
 
@@ -87,6 +126,29 @@ public class PaperMSPTCommand {
         return Arrays.asList(getColor(avgD), getColor(minD), getColor(maxD));
     }
 
+    // Octopus start
+    private static List<Component> eval(long[] times) {
+        if (times.length == 0) {
+            return Arrays.asList(getColor(0.0), getColor(0.0), getColor(0.0));
+        }
+        double sum = 0.0;
+        double min = Double.MAX_VALUE;
+        double max = Double.MIN_VALUE;
+        for (long time : times) {
+            double mspt = time * 1.0E-6D;
+            sum += mspt;
+            if (mspt < min) {
+                min = mspt;
+            }
+            if (mspt > max) {
+                max = mspt;
+            }
+        }
+        double avg = sum / times.length;
+        return Arrays.asList(getColor(avg), getColor(min), getColor(max));
+    }
+    // Octopus end
+
     private static Component getColor(double avg) {
         return text(DF.format(avg), avg >= 50 ? RED : avg >= 40 ? YELLOW : GREEN);
     }
