From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Fri, 25 Jul 2025 00:02:52 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 8704d9af920fe3f1e325841baa29e6cbe4ba6e4f..8f4958ee56db1b02b7aacaf0b636f7d7dec44cce 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -210,6 +210,13 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     }
     // Paper end - void damage configuration
 
+    // Octopus start - Get the MSPT for a single world
+    @Override
+    public @NotNull long[] getTickTimes() {
+        return world.getTickTimes();
+    }
+    // Octopus end - Get the MSPT for a single world
+
     // Paper start - Provide fast information methods
     @Override
     public int getEntityCount() {
