From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sat, 15 Mar 2025 21:20:46 +0800
Subject: [PATCH] Add clientConnected API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index cc4b2061ae9cbd5a30b297be477cde703203b3f5..3805b4c4b90e2989c3e4bd89f3154c68bf5e14f0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -216,6 +216,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private int hash = 0;
     private double health = 20;
     private boolean scaledHealth = false;
+    private boolean clientConnected = true;
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
@@ -3077,6 +3078,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return this.getHandle().allowsListing();
     }
 
+    // Octopus start
+    @Override
+    public boolean clientConnected() {
+        return clientConnected;
+    }
+
+    public void setClientConnected(boolean clientConnected) {
+        this.clientConnected = clientConnected;
+    }
+    // Octopus end
+
     // Paper start
     @Override
     public net.kyori.adventure.text.Component displayName() {
