From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Sun, 1 Feb 2026 15:54:02 +0800
Subject: [PATCH] Add PlayerMessagePreSendEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 55ef3217f39452b6fd158ddccbe92e5f41bee1b6..f1ec82f5b1dcdde9790d76f7ff5e407a7a44a4f2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2833,7 +2833,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     @Override
     public void sendMessage(final net.kyori.adventure.text.Component message) {
         if (getHandle().connection == null) return;
-        this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(message, false));
+        // Octopus start - Add PlayerMessagePreSendEvent
+        org.bukkit.event.player.PlayerMessagePreSendEvent event = new org.bukkit.event.player.PlayerMessagePreSendEvent(this, message);
+        if (event.callEvent()) {
+            this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(event.getComponent(), false));
+        }
+        // Octopus end - Add PlayerMessagePreSendEvent
     }
 
     @Override
@@ -3094,8 +3099,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         @Override
         public void sendMessage(net.md_5.bungee.api.ChatMessageType position, UUID sender, BaseComponent... components) {
             if (CraftPlayer.this.getHandle().connection == null) return;
-
-            CraftPlayer.this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(components, position == net.md_5.bungee.api.ChatMessageType.ACTION_BAR));
+            // Octopus start - Add PlayerMessagePreSendEvent
+            org.bukkit.event.player.PlayerMessagePreSendEvent event = new org.bukkit.event.player.PlayerMessagePreSendEvent(CraftPlayer.this, components);
+            if (event.callEvent()) {
+                CraftPlayer.this.getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(event.getComponent(), position == net.md_5.bungee.api.ChatMessageType.ACTION_BAR));
+            }
+            // Octopus end - Add PlayerMessagePreSendEvent
         }
 
         // Paper start
