From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Sun, 28 Sep 2025 00:03:13 +0800
Subject: [PATCH] Add cause for EntityTeleportEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index b38073628d3b1381ccc4e5c6c44b2b7ec8ba1273..49f453358f3e2752752c5aeddceaeda0bf68ac68 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -317,7 +317,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
         // Paper start - fix teleport event not being called
         org.bukkit.event.entity.EntityTeleportEvent event = new org.bukkit.event.entity.EntityTeleportEvent(
-            this, this.getLocation(), location);
+            this, this.getLocation(), location, cause); // Octopus - add cause
         // cancelling the event is handled differently for players and entities,
         // entities just stop teleporting, players will still teleport to the "from" location of the event
         if (!event.callEvent() || event.getTo() == null) {
