From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xiaozhangup <xiaozhangshiw@163.com>
Date: Tue, 14 Jul 2026 13:59:38 +0800
Subject: [PATCH] Custom world uuid in WorldCreator


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f12f863753c97c702f8507e955e1ad703c6e5942..e85b7b0277a053d4bd1e49badc92b8f3b3d0e220 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1219,6 +1219,14 @@ public final class CraftServer implements Server {
             dimensionKey,
             name
         );
+        if (creator.uuid() != null) {
+            loadedWorldData = new PaperWorldLoader.LoadedWorldData(
+                loadedWorldData.bukkitName(),
+                creator.uuid(),
+                loadedWorldData.pdc(),
+                loadedWorldData.levelOverrides()
+            );
+        }
         final PrimaryLevelData primaryLevelData = (PrimaryLevelData) this.console.getWorldData();
         WorldGenSettings worldGenSettings = LevelStorageSource.readExistingSavedData(this.console.storageSource, dimensionKey, registryAccess, WorldGenSettings.TYPE)
             .result()
