--- a/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -47,10 +_,17 @@
             ServerLevel level1 = level.getServer().getLevel(globalPos.dimension());
             if (level1 != null) {
                 PoiManager poiManager = level1.getPoiManager();
+                //Octopus start - Fix off world POI accessing
+                java.lang.Runnable scheduledRelease = () -> {
                 if (poiManager.exists(blockPos, holder -> true)) {
                     poiManager.release(blockPos);
                 }
-
+                };
+                if (!(level1.worldTaskQueueInstance.isHeldByAnyThread() && level1.worldTaskQueueInstance.isFullyHeldByCurrentThread()))
+                    level1.worldTaskQueueInstance.submitCallbackTask(scheduledRelease);
+                else
+                    scheduledRelease.run();
+                //Octopus end
                 DebugPackets.sendPoiTicketCountPacket(level, blockPos);
             }
         });
