From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Mon, 16 Sep 2024 21:45:35 +0800
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 7196594e07af19a14c320d77df893978525fe386..cc7ac83c263b5d11dda621c45c2b3641fbe3bd63 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -33,6 +33,8 @@ public interface ServerBuildInfo {
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
 
+    Key BRAND_OCTOPUS_ID = Key.key("happyland", "octopus");
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
