From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sat, 16 Nov 2024 21:41:18 +0800
Subject: [PATCH] Parallel world ticking


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index f5ead24639253ce051d632483252ac6c189ecfc2..bdba48e732756295c7b08a13aa2d95dbb4c7acb8 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4327,6 +4327,20 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public long[] getTickTimes();
 
+    /**
+     * Submit a task to be executed on the world thread.
+     * Usually after all other thread tasks have been executed.
+     * @param task the task to execute
+     */
+    void submitCallbackTask(Runnable task);
+
+    /**
+     * Submit a task to be executed on the world thread.
+     * Usually after all this world thread tasks have been executed.
+     * @param task the task to execute
+     */
+    void submitScopedTask(Runnable task);
+
     /**
      * Represents various map environment types that a world may be
      */
