From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sun, 9 Feb 2025 00:14:12 +0800
Subject: [PATCH] Add ComposterProductEvent


diff --git a/src/main/java/org/bukkit/event/block/ComposterProductEvent.java b/src/main/java/org/bukkit/event/block/ComposterProductEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..0619eb78edee5d701f35cc35706e30422188aca0
--- /dev/null
+++ b/src/main/java/org/bukkit/event/block/ComposterProductEvent.java
@@ -0,0 +1,36 @@
+package org.bukkit.event.block;
+
+import org.bukkit.block.Block;
+import org.bukkit.event.HandlerList;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public class ComposterProductEvent extends BlockEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private ItemStack itemStack;
+
+    public ComposterProductEvent(@NotNull final Block block, @NotNull final ItemStack itemStack) {
+        super(block);
+        this.itemStack = itemStack;
+    }
+
+    @NotNull
+    public ItemStack getItem() {
+        return itemStack;
+    }
+
+    public void setItem(@NotNull ItemStack itemStack) {
+        this.itemStack = itemStack;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
