From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Thu, 10 Oct 2024 23:52:38 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 551f2b2fc8f8830f35bb88045b30bee20424d2b1..705eca91dd6c04707814f763c8ab2af95f6d6c51 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -504,6 +504,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
     public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
     public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+
+    public long[] getTickTimes() {
+        return tickTimes5s.getTimes();
+    }
     // SparklyPaper end
 
     // Add env and gen to constructor, IWorldDataServer -> WorldDataServer
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 0921894dbc0d81b943827ecda6e6befdbfd715f5..48912de0c47007be3281e5b6c468d276d7f52e9f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2443,6 +2443,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return this.getStructures(x, z, struct -> registry.getKey(struct).equals(key));
     }
 
+    @Override
+    public @NotNull long[] getTickTimes() {
+        return world.getTickTimes();
+    }
+
     private List<GeneratedStructure> getStructures(int x, int z, Predicate<net.minecraft.world.level.levelgen.structure.Structure> predicate) {
         List<GeneratedStructure> structures = new ArrayList<>();
         for (StructureStart start : this.getHandle().structureManager().startsForStructure(new ChunkPos(x, z), predicate)) {
