From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Thu, 10 Oct 2024 23:52:38 +0800
Subject: [PATCH] Get the MSPT for a single world


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 35087b96b30765148592d4cb4fb6ec9094430fc6..1828077f050de17bb4be5a895bc4a511ffb125e9 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -504,6 +504,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
     public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
     public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+
+    public long[] getTickTimes() {
+        return tickTimes5s.getTimes();
+    }
     // SparklyPaper end
 
     // Add env and gen to constructor, IWorldDataServer -> WorldDataServer
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 40712284682f2805858be68fe9e5428b376a5943..ae34140f77ba293ce5fb53cb58d8a23496047639 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2400,6 +2400,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return this.getStructures(x, z, struct -> registry.getKey(struct).equals(key));
     }
 
+    @Override
+    public @NotNull long[] getTickTimes() {
+        return world.getTickTimes();
+    }
+
     private List<GeneratedStructure> getStructures(int x, int z, Predicate<net.minecraft.world.level.levelgen.structure.Structure> predicate) {
         List<GeneratedStructure> structures = new ArrayList<>();
         for (StructureStart start : this.getHandle().structureManager().startsForStructure(new ChunkPos(x, z), predicate)) {
