From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Zhang <xiaozhangshiw@163.com>
Date: Sun, 9 Feb 2025 00:14:40 +0800
Subject: [PATCH] Add ComposterProductEvent


diff --git a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
index f34159f8d6c51af2341bf49db0d6d6f0417919cf..0e21ea091e32b0b76aa029c367a1788b6fd76f5c 100644
--- a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
@@ -45,6 +45,7 @@ import net.minecraft.world.phys.shapes.VoxelShape;
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftBlockInventoryHolder;
 import org.bukkit.craftbukkit.util.DummyGeneratorAccess;
+import org.bukkit.event.block.ComposterProductEvent;
 // CraftBukkit end
 
 public class ComposterBlock extends Block implements WorldlyContainerHolder {
@@ -332,7 +333,9 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
         // CraftBukkit end
         if (!world.isClientSide) {
             Vec3 vec3d = Vec3.atLowerCornerWithOffset(pos, 0.5D, 1.01D, 0.5D).offsetRandom(world.random, 0.7F);
-            ItemEntity entityitem = new ItemEntity(world, vec3d.x(), vec3d.y(), vec3d.z(), new ItemStack(Items.BONE_MEAL));
+            ComposterProductEvent event = new ComposterProductEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), new ItemStack(Items.BONE_MEAL).asBukkitMirror());
+            event.callEvent();
+            ItemEntity entityitem = new ItemEntity(world, vec3d.x(), vec3d.y(), vec3d.z(), ItemStack.fromBukkitCopy(event.getItem()));
 
             entityitem.setDefaultPickUpDelay();
             world.addFreshEntity(entityitem);
@@ -433,7 +436,13 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
         int i = (Integer) state.getValue(ComposterBlock.LEVEL);
 
         // CraftBukkit - empty generatoraccess, blockposition
-        return (WorldlyContainer) (i == 8 ? new ComposterBlock.OutputContainer(state, world, pos, new ItemStack(Items.BONE_MEAL)) : (i < 7 ? new ComposterBlock.InputContainer(state, world, pos) : new ComposterBlock.EmptyContainer(world, pos)));
+        if (i == 8) {
+            ComposterProductEvent event = new ComposterProductEvent(world.getMinecraftWorld().getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), new ItemStack(Items.BONE_MEAL).asBukkitMirror());
+            event.callEvent();
+            return new ComposterBlock.OutputContainer(state, world, pos, ItemStack.fromBukkitCopy(event.getItem()));
+        } else {
+            return i < 7 ? new ComposterBlock.InputContainer(state, world, pos) : new ComposterBlock.EmptyContainer(world, pos);
+        }
     }
 
     public static class OutputContainer extends SimpleContainer implements WorldlyContainer {
@@ -468,7 +477,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
 
         @Override
         public boolean canTakeItemThroughFace(int slot, ItemStack stack, Direction dir) {
-            return !this.changed && dir == Direction.DOWN && stack.is(Items.BONE_MEAL);
+            return !this.changed && dir == Direction.DOWN && slot == 0;
         }
 
         @Override
